<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title></title>
  <style>
  @font-face {
  font-family: octicons-link;
  src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM+8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB/aFGpk3jaTY6xa8JAGMW/O62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v+k/0an2i+itHDw3v2+9+DBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3/I7AtxEJLtzzuZfI+VVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy/Lt7Kc+0vWY/gAgIIEqAN9we0pwKXreiMasxvabDQMM4riO+qxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw+ymhce7vwM9jSqO8JyVd5RH9gyTt2+J/yUmYlIR0s04n6+7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv/ocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi+W2+MjCzMIDApSwvXzC97Z4Ig8N/BxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh/8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT+AEjAwuDFpBmA9KMDEwMCh9i/v8H8sH0/4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9/lqYwOGZxeUelN2U2R6+cArgtCJpauW7UQBqnFkUsjAY/kOU1cP+DAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl+vvmM/byA48e6tWrKArm4ZJlCbdsrxksL1AwWn/yBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO//sdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd/89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF+9JOS0nbaaYDCQfwCJ7Au3AHj+LO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm+EBXuAbHmIMSRMs+4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL+hD7C1xoaHeLJSEao0FEW14ckxC+TU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13/+lm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl+9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O/AdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB///AA8AAQAAAAAAAAAAAAAAAAABAAAAAA==) format('woff');
}

.markdown-body .octicon {
  display: inline-block;
  fill: currentColor;
  vertical-align: text-bottom;
}

.markdown-body .anchor {
  float: left;
  line-height: 1;
  margin-left: -20px;
  padding-right: 4px;
}

.markdown-body .anchor:focus {
  outline: none;
}

.markdown-body h1 .octicon-link,
.markdown-body h2 .octicon-link,
.markdown-body h3 .octicon-link,
.markdown-body h4 .octicon-link,
.markdown-body h5 .octicon-link,
.markdown-body h6 .octicon-link {
  color: #1b1f23;
  vertical-align: middle;
  visibility: hidden;
}

.markdown-body h1:hover .anchor,
.markdown-body h2:hover .anchor,
.markdown-body h3:hover .anchor,
.markdown-body h4:hover .anchor,
.markdown-body h5:hover .anchor,
.markdown-body h6:hover .anchor {
  text-decoration: none;
}

.markdown-body h1:hover .anchor .octicon-link,
.markdown-body h2:hover .anchor .octicon-link,
.markdown-body h3:hover .anchor .octicon-link,
.markdown-body h4:hover .anchor .octicon-link,
.markdown-body h5:hover .anchor .octicon-link,
.markdown-body h6:hover .anchor .octicon-link {
  visibility: visible;
}

.markdown-body {
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  color: #24292e;
  line-height: 1.5;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;
  font-size: 16px;
  line-height: 1.5;
  word-wrap: break-word;
}

.markdown-body .pl-c {
  color: #6a737d;
}

.markdown-body .pl-c1,
.markdown-body .pl-s .pl-v {
  color: #005cc5;
}

.markdown-body .pl-e,
.markdown-body .pl-en {
  color: #6f42c1;
}

.markdown-body .pl-s .pl-s1,
.markdown-body .pl-smi {
  color: #24292e;
}

.markdown-body .pl-ent {
  color: #22863a;
}

.markdown-body .pl-k {
  color: #d73a49;
}

.markdown-body .pl-pds,
.markdown-body .pl-s,
.markdown-body .pl-s .pl-pse .pl-s1,
.markdown-body .pl-sr,
.markdown-body .pl-sr .pl-cce,
.markdown-body .pl-sr .pl-sra,
.markdown-body .pl-sr .pl-sre {
  color: #032f62;
}

.markdown-body .pl-smw,
.markdown-body .pl-v {
  color: #e36209;
}

.markdown-body .pl-bu {
  color: #b31d28;
}

.markdown-body .pl-ii {
  background-color: #b31d28;
  color: #fafbfc;
}

.markdown-body .pl-c2 {
  background-color: #d73a49;
  color: #fafbfc;
}

.markdown-body .pl-c2:before {
  content: "^M";
}

.markdown-body .pl-sr .pl-cce {
  color: #22863a;
  font-weight: 700;
}

.markdown-body .pl-ml {
  color: #735c0f;
}

.markdown-body .pl-mh,
.markdown-body .pl-mh .pl-en,
.markdown-body .pl-ms {
  color: #005cc5;
  font-weight: 700;
}

.markdown-body .pl-mi {
  color: #24292e;
  font-style: italic;
}

.markdown-body .pl-mb {
  color: #24292e;
  font-weight: 700;
}

.markdown-body .pl-md {
  background-color: #ffeef0;
  color: #b31d28;
}

.markdown-body .pl-mi1 {
  background-color: #f0fff4;
  color: #22863a;
}

.markdown-body .pl-mc {
  background-color: #ffebda;
  color: #e36209;
}

.markdown-body .pl-mi2 {
  background-color: #005cc5;
  color: #f6f8fa;
}

.markdown-body .pl-mdr {
  color: #6f42c1;
  font-weight: 700;
}

.markdown-body .pl-ba {
  color: #586069;
}

.markdown-body .pl-sg {
  color: #959da5;
}

.markdown-body .pl-corl {
  color: #032f62;
  text-decoration: underline;
}

.markdown-body details {
  display: block;
}

.markdown-body summary {
  display: list-item;
}

.markdown-body a {
  background-color: transparent;
}

.markdown-body a:active,
.markdown-body a:hover {
  outline-width: 0;
}

.markdown-body strong {
  font-weight: inherit;
  font-weight: bolder;
}

.markdown-body h1 {
  font-size: 2em;
  margin: .67em 0;
}

.markdown-body img {
  border-style: none;
}

.markdown-body code,
.markdown-body kbd,
.markdown-body pre {
  font-family: monospace,monospace;
  font-size: 1em;
}

.markdown-body hr {
  box-sizing: content-box;
  height: 0;
  overflow: visible;
}

.markdown-body input {
  font: inherit;
  margin: 0;
}

.markdown-body input {
  overflow: visible;
}

.markdown-body [type=checkbox] {
  box-sizing: border-box;
  padding: 0;
}

.markdown-body * {
  box-sizing: border-box;
}

.markdown-body input {
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
}

.markdown-body a {
  color: #0366d6;
  text-decoration: none;
}

.markdown-body a:hover {
  text-decoration: underline;
}

.markdown-body strong {
  font-weight: 600;
}

.markdown-body hr {
  background: transparent;
  border: 0;
  border-bottom: 1px solid #dfe2e5;
  height: 0;
  margin: 15px 0;
  overflow: hidden;
}

.markdown-body hr:before {
  content: "";
  display: table;
}

.markdown-body hr:after {
  clear: both;
  content: "";
  display: table;
}

.markdown-body table {
  border-collapse: collapse;
  border-spacing: 0;
}

.markdown-body td,
.markdown-body th {
  padding: 0;
}

.markdown-body details summary {
  cursor: pointer;
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  margin-bottom: 0;
  margin-top: 0;
}

.markdown-body h1 {
  font-size: 32px;
}

.markdown-body h1,
.markdown-body h2 {
  font-weight: 600;
}

.markdown-body h2 {
  font-size: 24px;
}

.markdown-body h3 {
  font-size: 20px;
}

.markdown-body h3,
.markdown-body h4 {
  font-weight: 600;
}

.markdown-body h4 {
  font-size: 16px;
}

.markdown-body h5 {
  font-size: 14px;
}

.markdown-body h5,
.markdown-body h6 {
  font-weight: 600;
}

.markdown-body h6 {
  font-size: 12px;
}

.markdown-body p {
  margin-bottom: 10px;
  margin-top: 0;
}

.markdown-body blockquote {
  margin: 0;
}

.markdown-body ol,
.markdown-body ul {
  margin-bottom: 0;
  margin-top: 0;
  padding-left: 0;
}

.markdown-body ol ol,
.markdown-body ul ol {
  list-style-type: lower-roman;
}

.markdown-body ol ol ol,
.markdown-body ol ul ol,
.markdown-body ul ol ol,
.markdown-body ul ul ol {
  list-style-type: lower-alpha;
}

.markdown-body dd {
  margin-left: 0;
}

.markdown-body code,
.markdown-body pre {
  font-family: SFMono-Regular,Consolas,Liberation Mono,Menlo,Courier,monospace;
  font-size: 12px;
}

.markdown-body pre {
  margin-bottom: 0;
  margin-top: 0;
}

.markdown-body input::-webkit-inner-spin-button,
.markdown-body input::-webkit-outer-spin-button {
  -webkit-appearance: none;
  appearance: none;
  margin: 0;
}

.markdown-body .border {
  border: 1px solid #e1e4e8!important;
}

.markdown-body .border-0 {
  border: 0!important;
}

.markdown-body .border-bottom {
  border-bottom: 1px solid #e1e4e8!important;
}

.markdown-body .rounded-1 {
  border-radius: 3px!important;
}

.markdown-body .bg-white {
  background-color: #fff!important;
}

.markdown-body .bg-gray-light {
  background-color: #fafbfc!important;
}

.markdown-body .text-gray-light {
  color: #6a737d!important;
}

.markdown-body .mb-0 {
  margin-bottom: 0!important;
}

.markdown-body .my-2 {
  margin-bottom: 8px!important;
  margin-top: 8px!important;
}

.markdown-body .pl-0 {
  padding-left: 0!important;
}

.markdown-body .py-0 {
  padding-bottom: 0!important;
  padding-top: 0!important;
}

.markdown-body .pl-1 {
  padding-left: 4px!important;
}

.markdown-body .pl-2 {
  padding-left: 8px!important;
}

.markdown-body .py-2 {
  padding-bottom: 8px!important;
  padding-top: 8px!important;
}

.markdown-body .pl-3,
.markdown-body .px-3 {
  padding-left: 16px!important;
}

.markdown-body .px-3 {
  padding-right: 16px!important;
}

.markdown-body .pl-4 {
  padding-left: 24px!important;
}

.markdown-body .pl-5 {
  padding-left: 32px!important;
}

.markdown-body .pl-6 {
  padding-left: 40px!important;
}

.markdown-body .f6 {
  font-size: 12px!important;
}

.markdown-body .lh-condensed {
  line-height: 1.25!important;
}

.markdown-body .text-bold {
  font-weight: 600!important;
}

.markdown-body:before {
  content: "";
  display: table;
}

.markdown-body:after {
  clear: both;
  content: "";
  display: table;
}

.markdown-body>:first-child {
  margin-top: 0!important;
}

.markdown-body>:last-child {
  margin-bottom: 0!important;
}

.markdown-body a:not([href]) {
  color: inherit;
  text-decoration: none;
}

.markdown-body blockquote,
.markdown-body dl,
.markdown-body ol,
.markdown-body p,
.markdown-body pre,
.markdown-body table,
.markdown-body ul {
  margin-bottom: 16px;
  margin-top: 0;
}

.markdown-body hr {
  background-color: #e1e4e8;
  border: 0;
  height: .25em;
  margin: 24px 0;
  padding: 0;
}

.markdown-body blockquote {
  border-left: .25em solid #dfe2e5;
  color: #6a737d;
  padding: 0 1em;
}

.markdown-body blockquote>:first-child {
  margin-top: 0;
}

.markdown-body blockquote>:last-child {
  margin-bottom: 0;
}

.markdown-body kbd {
  background-color: #fafbfc;
  border: 1px solid #c6cbd1;
  border-bottom-color: #959da5;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #959da5;
  color: #444d56;
  display: inline-block;
  font-size: 11px;
  line-height: 10px;
  padding: 3px 5px;
  vertical-align: middle;
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  font-weight: 600;
  line-height: 1.25;
  margin-bottom: 16px;
  margin-top: 24px;
}

.markdown-body h1 {
  font-size: 2em;
}

.markdown-body h1,
.markdown-body h2 {
  border-bottom: 1px solid #eaecef;
  padding-bottom: .3em;
}

.markdown-body h2 {
  font-size: 1.5em;
}

.markdown-body h3 {
  font-size: 1.25em;
}

.markdown-body h4 {
  font-size: 1em;
}

.markdown-body h5 {
  font-size: .875em;
}

.markdown-body h6 {
  color: #6a737d;
  font-size: .85em;
}

.markdown-body ol,
.markdown-body ul {
  padding-left: 2em;
}

.markdown-body ol ol,
.markdown-body ol ul,
.markdown-body ul ol,
.markdown-body ul ul {
  margin-bottom: 0;
  margin-top: 0;
}

.markdown-body li {
  word-wrap: break-all;
}

.markdown-body li>p {
  margin-top: 16px;
}

.markdown-body li+li {
  margin-top: .25em;
}

.markdown-body dl {
  padding: 0;
}

.markdown-body dl dt {
  font-size: 1em;
  font-style: italic;
  font-weight: 600;
  margin-top: 16px;
  padding: 0;
}

.markdown-body dl dd {
  margin-bottom: 16px;
  padding: 0 16px;
}

.markdown-body table {
  display: block;
  overflow: auto;
  width: 100%;
}

.markdown-body table th {
  font-weight: 600;
}

.markdown-body table td,
.markdown-body table th {
  border: 1px solid #dfe2e5;
  padding: 6px 13px;
}

.markdown-body table tr {
  background-color: #fff;
  border-top: 1px solid #c6cbd1;
}

.markdown-body table tr:nth-child(2n) {
  background-color: #f6f8fa;
}

.markdown-body img {
  background-color: #fff;
  box-sizing: content-box;
  max-width: 100%;
}

.markdown-body img[align=right] {
  padding-left: 20px;
}

.markdown-body img[align=left] {
  padding-right: 20px;
}

.markdown-body code {
  background-color: rgba(27,31,35,.05);
  border-radius: 3px;
  font-size: 85%;
  margin: 0;
  padding: .2em .4em;
}

.markdown-body pre {
  word-wrap: normal;
}

.markdown-body pre>code {
  background: transparent;
  border: 0;
  font-size: 100%;
  margin: 0;
  padding: 0;
  white-space: pre;
  word-break: normal;
}

.markdown-body .highlight {
  margin-bottom: 16px;
}

.markdown-body .highlight pre {
  margin-bottom: 0;
  word-break: normal;
}

.markdown-body .highlight pre,
.markdown-body pre {
  background-color: #f6f8fa;
  border-radius: 3px;
  font-size: 85%;
  line-height: 1.45;
  overflow: auto;
  padding: 16px;
}

.markdown-body pre code {
  background-color: transparent;
  border: 0;
  display: inline;
  line-height: inherit;
  margin: 0;
  max-width: auto;
  overflow: visible;
  padding: 0;
  word-wrap: normal;
}

.markdown-body .commit-tease-sha {
  color: #444d56;
  display: inline-block;
  font-family: SFMono-Regular,Consolas,Liberation Mono,Menlo,Courier,monospace;
  font-size: 90%;
}

.markdown-body .blob-wrapper {
  border-bottom-left-radius: 3px;
  border-bottom-right-radius: 3px;
  overflow-x: auto;
  overflow-y: hidden;
}

.markdown-body .blob-wrapper-embedded {
  max-height: 240px;
  overflow-y: auto;
}

.markdown-body .blob-num {
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  color: rgba(27,31,35,.3);
  cursor: pointer;
  font-family: SFMono-Regular,Consolas,Liberation Mono,Menlo,Courier,monospace;
  font-size: 12px;
  line-height: 20px;
  min-width: 50px;
  padding-left: 10px;
  padding-right: 10px;
  text-align: right;
  user-select: none;
  vertical-align: top;
  white-space: nowrap;
  width: 1%;
}

.markdown-body .blob-num:hover {
  color: rgba(27,31,35,.6);
}

.markdown-body .blob-num:before {
  content: attr(data-line-number);
}

.markdown-body .blob-code {
  line-height: 20px;
  padding-left: 10px;
  padding-right: 10px;
  position: relative;
  vertical-align: top;
}

.markdown-body .blob-code-inner {
  color: #24292e;
  font-family: SFMono-Regular,Consolas,Liberation Mono,Menlo,Courier,monospace;
  font-size: 12px;
  overflow: visible;
  white-space: pre;
  word-wrap: normal;
}

.markdown-body .pl-token.active,
.markdown-body .pl-token:hover {
  background: #ffea7f;
  cursor: pointer;
}

.markdown-body kbd {
  background-color: #fafbfc;
  border: 1px solid #d1d5da;
  border-bottom-color: #c6cbd1;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #c6cbd1;
  color: #444d56;
  display: inline-block;
  font: 11px SFMono-Regular,Consolas,Liberation Mono,Menlo,Courier,monospace;
  line-height: 10px;
  padding: 3px 5px;
  vertical-align: middle;
}

.markdown-body :checked+.radio-label {
  border-color: #0366d6;
  position: relative;
  z-index: 1;
}

.markdown-body .tab-size[data-tab-size="1"] {
  -moz-tab-size: 1;
  tab-size: 1;
}

.markdown-body .tab-size[data-tab-size="2"] {
  -moz-tab-size: 2;
  tab-size: 2;
}

.markdown-body .tab-size[data-tab-size="3"] {
  -moz-tab-size: 3;
  tab-size: 3;
}

.markdown-body .tab-size[data-tab-size="4"] {
  -moz-tab-size: 4;
  tab-size: 4;
}

.markdown-body .tab-size[data-tab-size="5"] {
  -moz-tab-size: 5;
  tab-size: 5;
}

.markdown-body .tab-size[data-tab-size="6"] {
  -moz-tab-size: 6;
  tab-size: 6;
}

.markdown-body .tab-size[data-tab-size="7"] {
  -moz-tab-size: 7;
  tab-size: 7;
}

.markdown-body .tab-size[data-tab-size="8"] {
  -moz-tab-size: 8;
  tab-size: 8;
}

.markdown-body .tab-size[data-tab-size="9"] {
  -moz-tab-size: 9;
  tab-size: 9;
}

.markdown-body .tab-size[data-tab-size="10"] {
  -moz-tab-size: 10;
  tab-size: 10;
}

.markdown-body .tab-size[data-tab-size="11"] {
  -moz-tab-size: 11;
  tab-size: 11;
}

.markdown-body .tab-size[data-tab-size="12"] {
  -moz-tab-size: 12;
  tab-size: 12;
}

.markdown-body .task-list-item {
  list-style-type: none;
}

.markdown-body .task-list-item+.task-list-item {
  margin-top: 3px;
}

.markdown-body .task-list-item input {
  margin: 0 .2em .25em -1.6em;
  vertical-align: middle;
}

.markdown-body hr {
  border-bottom-color: #eee;
}

.markdown-body .pl-0 {
  padding-left: 0!important;
}

.markdown-body .pl-1 {
  padding-left: 4px!important;
}

.markdown-body .pl-2 {
  padding-left: 8px!important;
}

.markdown-body .pl-3 {
  padding-left: 16px!important;
}

.markdown-body .pl-4 {
  padding-left: 24px!important;
}

.markdown-body .pl-5 {
  padding-left: 32px!important;
}

.markdown-body .pl-6 {
  padding-left: 40px!important;
}

.markdown-body .pl-7 {
  padding-left: 48px!important;
}

.markdown-body .pl-8 {
  padding-left: 64px!important;
}

.markdown-body .pl-9 {
  padding-left: 80px!important;
}

.markdown-body .pl-10 {
  padding-left: 96px!important;
}

.markdown-body .pl-11 {
  padding-left: 112px!important;
}

.markdown-body .pl-12 {
  padding-left: 128px!important;
}

/*# sourceURL=webpack://./node_modules/github-markdown-css/github-markdown.css */
/*# sourceMappingURL=data:application/json;charset=utf-8;base64,{"version":3,"sources":["webpack://./node_modules/github-markdown-css/github-markdown.css"],"names":[],"mappings":"AAAA;EACE,0BAA0B;EAC1B,2DAAqtE;AACvtE;;AAEA;EACE,qBAAqB;EACrB,kBAAkB;EAClB,2BAA2B;AAC7B;;AAEA;EACE,WAAW;EACX,cAAc;EACd,kBAAkB;EAClB,kBAAkB;AACpB;;AAEA;EACE,aAAa;AACf;;AAEA;;;;;;EAME,cAAc;EACd,sBAAsB;EACtB,kBAAkB;AACpB;;AAEA;;;;;;EAME,qBAAqB;AACvB;;AAEA;;;;;;EAME,mBAAmB;AACrB;;AAEA;EACE,0BAA0B;EAC1B,8BAA8B;EAC9B,cAAc;EACd,gBAAgB;EAChB,kIAAkI;EAClI,eAAe;EACf,gBAAgB;EAChB,qBAAqB;AACvB;;AAEA;EACE,cAAc;AAChB;;AAEA;;EAEE,cAAc;AAChB;;AAEA;;EAEE,cAAc;AAChB;;AAEA;;EAEE,cAAc;AAChB;;AAEA;EACE,cAAc;AAChB;;AAEA;EACE,cAAc;AAChB;;AAEA;;;;;;;EAOE,cAAc;AAChB;;AAEA;;EAEE,cAAc;AAChB;;AAEA;EACE,cAAc;AAChB;;AAEA;EACE,yBAAyB;EACzB,cAAc;AAChB;;AAEA;EACE,yBAAyB;EACzB,cAAc;AAChB;;AAEA;EACE,aAAa;AACf;;AAEA;EACE,cAAc;EACd,gBAAgB;AAClB;;AAEA;EACE,cAAc;AAChB;;AAEA;;;EAGE,cAAc;EACd,gBAAgB;AAClB;;AAEA;EACE,cAAc;EACd,kBAAkB;AACpB;;AAEA;EACE,cAAc;EACd,gBAAgB;AAClB;;AAEA;EACE,yBAAyB;EACzB,cAAc;AAChB;;AAEA;EACE,yBAAyB;EACzB,cAAc;AAChB;;AAEA;EACE,yBAAyB;EACzB,cAAc;AAChB;;AAEA;EACE,yBAAyB;EACzB,cAAc;AAChB;;AAEA;EACE,cAAc;EACd,gBAAgB;AAClB;;AAEA;EACE,cAAc;AAChB;;AAEA;EACE,cAAc;AAChB;;AAEA;EACE,cAAc;EACd,0BAA0B;AAC5B;;AAEA;EACE,cAAc;AAChB;;AAEA;EACE,kBAAkB;AACpB;;AAEA;EACE,6BAA6B;AAC/B;;AAEA;;EAEE,gBAAgB;AAClB;;AAEA;EACE,oBAAoB;EACpB,mBAAmB;AACrB;;AAEA;EACE,cAAc;EACd,eAAe;AACjB;;AAEA;EACE,kBAAkB;AACpB;;AAEA;;;EAGE,gCAAgC;EAChC,cAAc;AAChB;;AAEA;EACE,uBAAuB;EACvB,SAAS;EACT,iBAAiB;AACnB;;AAEA;EACE,aAAa;EACb,SAAS;AACX;;AAEA;EACE,iBAAiB;AACnB;;AAEA;EACE,sBAAsB;EACtB,UAAU;AACZ;;AAEA;EACE,sBAAsB;AACxB;;AAEA;EACE,oBAAoB;EACpB,kBAAkB;EAClB,oBAAoB;AACtB;;AAEA;EACE,cAAc;EACd,qBAAqB;AACvB;;AAEA;EACE,0BAA0B;AAC5B;;AAEA;EACE,gBAAgB;AAClB;;AAEA;EACE,uBAAuB;EACvB,SAAS;EACT,gCAAgC;EAChC,SAAS;EACT,cAAc;EACd,gBAAgB;AAClB;;AAEA;EACE,WAAW;EACX,cAAc;AAChB;;AAEA;EACE,WAAW;EACX,WAAW;EACX,cAAc;AAChB;;AAEA;EACE,yBAAyB;EACzB,iBAAiB;AACnB;;AAEA;;EAEE,UAAU;AACZ;;AAEA;EACE,eAAe;AACjB;;AAEA;;;;;;EAME,gBAAgB;EAChB,aAAa;AACf;;AAEA;EACE,eAAe;AACjB;;AAEA;;EAEE,gBAAgB;AAClB;;AAEA;EACE,eAAe;AACjB;;AAEA;EACE,eAAe;AACjB;;AAEA;;EAEE,gBAAgB;AAClB;;AAEA;EACE,eAAe;AACjB;;AAEA;EACE,eAAe;AACjB;;AAEA;;EAEE,gBAAgB;AAClB;;AAEA;EACE,eAAe;AACjB;;AAEA;EACE,mBAAmB;EACnB,aAAa;AACf;;AAEA;EACE,SAAS;AACX;;AAEA;;EAEE,gBAAgB;EAChB,aAAa;EACb,eAAe;AACjB;;AAEA;;EAEE,4BAA4B;AAC9B;;AAEA;;;;EAIE,4BAA4B;AAC9B;;AAEA;EACE,cAAc;AAChB;;AAEA;;EAEE,4EAA4E;EAC5E,eAAe;AACjB;;AAEA;EACE,gBAAgB;EAChB,aAAa;AACf;;AAEA;;EAEE,wBAAwB;EACxB,gBAAgB;EAChB,SAAS;AACX;;AAEA;EACE,mCAAmC;AACrC;;AAEA;EACE,mBAAmB;AACrB;;AAEA;EACE,0CAA0C;AAC5C;;AAEA;EACE,4BAA4B;AAC9B;;AAEA;EACE,gCAAgC;AAClC;;AAEA;EACE,mCAAmC;AACrC;;AAEA;EACE,wBAAwB;AAC1B;;AAEA;EACE,0BAA0B;AAC5B;;AAEA;EACE,4BAA4B;EAC5B,yBAAyB;AAC3B;;AAEA;EACE,yBAAyB;AAC3B;;AAEA;EACE,2BAA2B;EAC3B,wBAAwB;AAC1B;;AAEA;EACE,2BAA2B;AAC7B;;AAEA;EACE,2BAA2B;AAC7B;;AAEA;EACE,6BAA6B;EAC7B,0BAA0B;AAC5B;;AAEA;;EAEE,4BAA4B;AAC9B;;AAEA;EACE,6BAA6B;AAC/B;;AAEA;EACE,4BAA4B;AAC9B;;AAEA;EACE,4BAA4B;AAC9B;;AAEA;EACE,4BAA4B;AAC9B;;AAEA;EACE,yBAAyB;AAC3B;;AAEA;EACE,2BAA2B;AAC7B;;AAEA;EACE,0BAA0B;AAC5B;;AAEA;EACE,WAAW;EACX,cAAc;AAChB;;AAEA;EACE,WAAW;EACX,WAAW;EACX,cAAc;AAChB;;AAEA;EACE,uBAAuB;AACzB;;AAEA;EACE,0BAA0B;AAC5B;;AAEA;EACE,cAAc;EACd,qBAAqB;AACvB;;AAEA;;;;;;;EAOE,mBAAmB;EACnB,aAAa;AACf;;AAEA;EACE,yBAAyB;EACzB,SAAS;EACT,aAAa;EACb,cAAc;EACd,UAAU;AACZ;;AAEA;EACE,gCAAgC;EAChC,cAAc;EACd,cAAc;AAChB;;AAEA;EACE,aAAa;AACf;;AAEA;EACE,gBAAgB;AAClB;;AAEA;EACE,yBAAyB;EACzB,yBAAyB;EACzB,4BAA4B;EAC5B,kBAAkB;EAClB,kCAAkC;EAClC,cAAc;EACd,qBAAqB;EACrB,eAAe;EACf,iBAAiB;EACjB,gBAAgB;EAChB,sBAAsB;AACxB;;AAEA;;;;;;EAME,gBAAgB;EAChB,iBAAiB;EACjB,mBAAmB;EACnB,gBAAgB;AAClB;;AAEA;EACE,cAAc;AAChB;;AAEA;;EAEE,gCAAgC;EAChC,oBAAoB;AACtB;;AAEA;EACE,gBAAgB;AAClB;;AAEA;EACE,iBAAiB;AACnB;;AAEA;EACE,cAAc;AAChB;;AAEA;EACE,iBAAiB;AACnB;;AAEA;EACE,cAAc;EACd,gBAAgB;AAClB;;AAEA;;EAEE,iBAAiB;AACnB;;AAEA;;;;EAIE,gBAAgB;EAChB,aAAa;AACf;;AAEA;EACE,oBAAoB;AACtB;;AAEA;EACE,gBAAgB;AAClB;;AAEA;EACE,iBAAiB;AACnB;;AAEA;EACE,UAAU;AACZ;;AAEA;EACE,cAAc;EACd,kBAAkB;EAClB,gBAAgB;EAChB,gBAAgB;EAChB,UAAU;AACZ;;AAEA;EACE,mBAAmB;EACnB,eAAe;AACjB;;AAEA;EACE,cAAc;EACd,cAAc;EACd,WAAW;AACb;;AAEA;EACE,gBAAgB;AAClB;;AAEA;;EAEE,yBAAyB;EACzB,iBAAiB;AACnB;;AAEA;EACE,sBAAsB;EACtB,6BAA6B;AAC/B;;AAEA;EACE,yBAAyB;AAC3B;;AAEA;EACE,sBAAsB;EACtB,uBAAuB;EACvB,eAAe;AACjB;;AAEA;EACE,kBAAkB;AACpB;;AAEA;EACE,mBAAmB;AACrB;;AAEA;EACE,oCAAoC;EACpC,kBAAkB;EAClB,cAAc;EACd,SAAS;EACT,kBAAkB;AACpB;;AAEA;EACE,iBAAiB;AACnB;;AAEA;EACE,uBAAuB;EACvB,SAAS;EACT,eAAe;EACf,SAAS;EACT,UAAU;EACV,gBAAgB;EAChB,kBAAkB;AACpB;;AAEA;EACE,mBAAmB;AACrB;;AAEA;EACE,gBAAgB;EAChB,kBAAkB;AACpB;;AAEA;;EAEE,yBAAyB;EACzB,kBAAkB;EAClB,cAAc;EACd,iBAAiB;EACjB,cAAc;EACd,aAAa;AACf;;AAEA;EACE,6BAA6B;EAC7B,SAAS;EACT,eAAe;EACf,oBAAoB;EACpB,SAAS;EACT,eAAe;EACf,iBAAiB;EACjB,UAAU;EACV,iBAAiB;AACnB;;AAEA;EACE,cAAc;EACd,qBAAqB;EACrB,4EAA4E;EAC5E,cAAc;AAChB;;AAEA;EACE,8BAA8B;EAC9B,+BAA+B;EAC/B,gBAAgB;EAChB,kBAAkB;AACpB;;AAEA;EACE,iBAAiB;EACjB,gBAAgB;AAClB;;AAEA;EACE,sBAAsB;EACtB,qBAAqB;EACrB,yBAAyB;EACzB,wBAAwB;EACxB,eAAe;EACf,4EAA4E;EAC5E,eAAe;EACf,iBAAiB;EACjB,eAAe;EACf,kBAAkB;EAClB,mBAAmB;EACnB,iBAAiB;EACjB,iBAAiB;EACjB,mBAAmB;EACnB,mBAAmB;EACnB,SAAS;AACX;;AAEA;EACE,wBAAwB;AAC1B;;AAEA;EACE,+BAA+B;AACjC;;AAEA;EACE,iBAAiB;EACjB,kBAAkB;EAClB,mBAAmB;EACnB,kBAAkB;EAClB,mBAAmB;AACrB;;AAEA;EACE,cAAc;EACd,4EAA4E;EAC5E,eAAe;EACf,iBAAiB;EACjB,gBAAgB;EAChB,iBAAiB;AACnB;;AAEA;;EAEE,mBAAmB;EACnB,eAAe;AACjB;;AAEA;EACE,yBAAyB;EACzB,yBAAyB;EACzB,4BAA4B;EAC5B,kBAAkB;EAClB,kCAAkC;EAClC,cAAc;EACd,qBAAqB;EACrB,0EAA0E;EAC1E,iBAAiB;EACjB,gBAAgB;EAChB,sBAAsB;AACxB;;AAEA;EACE,qBAAqB;EACrB,kBAAkB;EAClB,UAAU;AACZ;;AAEA;EACE,gBAAgB;EAChB,WAAW;AACb;;AAEA;EACE,gBAAgB;EAChB,WAAW;AACb;;AAEA;EACE,gBAAgB;EAChB,WAAW;AACb;;AAEA;EACE,gBAAgB;EAChB,WAAW;AACb;;AAEA;EACE,gBAAgB;EAChB,WAAW;AACb;;AAEA;EACE,gBAAgB;EAChB,WAAW;AACb;;AAEA;EACE,gBAAgB;EAChB,WAAW;AACb;;AAEA;EACE,gBAAgB;EAChB,WAAW;AACb;;AAEA;EACE,gBAAgB;EAChB,WAAW;AACb;;AAEA;EACE,iBAAiB;EACjB,YAAY;AACd;;AAEA;EACE,iBAAiB;EACjB,YAAY;AACd;;AAEA;EACE,iBAAiB;EACjB,YAAY;AACd;;AAEA;EACE,qBAAqB;AACvB;;AAEA;EACE,eAAe;AACjB;;AAEA;EACE,2BAA2B;EAC3B,sBAAsB;AACxB;;AAEA;EACE,yBAAyB;AAC3B;;AAEA;EACE,yBAAyB;AAC3B;;AAEA;EACE,2BAA2B;AAC7B;;AAEA;EACE,2BAA2B;AAC7B;;AAEA;EACE,4BAA4B;AAC9B;;AAEA;EACE,4BAA4B;AAC9B;;AAEA;EACE,4BAA4B;AAC9B;;AAEA;EACE,4BAA4B;AAC9B;;AAEA;EACE,4BAA4B;AAC9B;;AAEA;EACE,4BAA4B;AAC9B;;AAEA;EACE,4BAA4B;AAC9B;;AAEA;EACE,4BAA4B;AAC9B;;AAEA;EACE,6BAA6B;AAC/B;;AAEA;EACE,6BAA6B;AAC/B","sourceRoot":""} */
  </style>
  <style>
  /**
 * prism.js default theme for JavaScript, CSS and HTML
 * Based on dabblet (http://dabblet.com)
 * @author Lea Verou
 */

code[class*="language-"],
pre[class*="language-"] {
	color: black;
	background: none;
	text-shadow: 0 1px white;
	font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
	font-size: 1em;
	text-align: left;
	white-space: pre;
	word-spacing: normal;
	word-break: normal;
	word-wrap: normal;
	line-height: 1.5;

	-moz-tab-size: 4;
	-o-tab-size: 4;
	tab-size: 4;

	-webkit-hyphens: none;
	-moz-hyphens: none;
	-ms-hyphens: none;
	hyphens: none;
}

pre[class*="language-"]::-moz-selection, pre[class*="language-"] ::-moz-selection,
code[class*="language-"]::-moz-selection, code[class*="language-"] ::-moz-selection {
	text-shadow: none;
	background: #b3d4fc;
}

pre[class*="language-"]::selection, pre[class*="language-"] ::selection,
code[class*="language-"]::selection, code[class*="language-"] ::selection {
	text-shadow: none;
	background: #b3d4fc;
}

@media print {
	code[class*="language-"],
	pre[class*="language-"] {
		text-shadow: none;
	}
}

/* Code blocks */
pre[class*="language-"] {
	padding: 1em;
	margin: .5em 0;
	overflow: auto;
}

:not(pre) > code[class*="language-"],
pre[class*="language-"] {
	background: #f5f2f0;
}

/* Inline code */
:not(pre) > code[class*="language-"] {
	padding: .1em;
	border-radius: .3em;
	white-space: normal;
}

.token.comment,
.token.prolog,
.token.doctype,
.token.cdata {
	color: slategray;
}

.token.punctuation {
	color: #999;
}

.token.namespace {
	opacity: .7;
}

.token.property,
.token.tag,
.token.boolean,
.token.number,
.token.constant,
.token.symbol,
.token.deleted {
	color: #905;
}

.token.selector,
.token.attr-name,
.token.string,
.token.char,
.token.builtin,
.token.inserted {
	color: #690;
}

.token.operator,
.token.entity,
.token.url,
.language-css .token.string,
.style .token.string {
	color: #9a6e3a;
	/* This background color was intended by the author of this theme. */
	background: hsla(0, 0%, 100%, .5);
}

.token.atrule,
.token.attr-value,
.token.keyword {
	color: #07a;
}

.token.function,
.token.class-name {
	color: #DD4A68;
}

.token.regex,
.token.important,
.token.variable {
	color: #e90;
}

.token.important,
.token.bold {
	font-weight: bold;
}
.token.italic {
	font-style: italic;
}

.token.entity {
	cursor: help;
}

/*# sourceURL=webpack://./node_modules/prismjs/themes/prism.css */
/*# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIndlYnBhY2s6Ly8uL25vZGVfbW9kdWxlcy9wcmlzbWpzL3RoZW1lcy9wcmlzbS5jc3MiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUE7Ozs7RUFJRTs7QUFFRjs7Q0FFQyxZQUFZO0NBQ1osZ0JBQWdCO0NBQ2hCLHdCQUF3QjtDQUN4QixzRUFBc0U7Q0FDdEUsY0FBYztDQUNkLGdCQUFnQjtDQUNoQixnQkFBZ0I7Q0FDaEIsb0JBQW9CO0NBQ3BCLGtCQUFrQjtDQUNsQixpQkFBaUI7Q0FDakIsZ0JBQWdCOztDQUVoQixnQkFBZ0I7Q0FDaEIsY0FBYztDQUNkLFdBQVc7O0NBRVgscUJBQXFCO0NBQ3JCLGtCQUFrQjtDQUNsQixpQkFBaUI7Q0FDakIsYUFBYTtBQUNkOztBQUVBOztDQUVDLGlCQUFpQjtDQUNqQixtQkFBbUI7QUFDcEI7O0FBRUE7O0NBRUMsaUJBQWlCO0NBQ2pCLG1CQUFtQjtBQUNwQjs7QUFFQTtDQUNDOztFQUVDLGlCQUFpQjtDQUNsQjtBQUNEOztBQUVBLGdCQUFnQjtBQUNoQjtDQUNDLFlBQVk7Q0FDWixjQUFjO0NBQ2QsY0FBYztBQUNmOztBQUVBOztDQUVDLG1CQUFtQjtBQUNwQjs7QUFFQSxnQkFBZ0I7QUFDaEI7Q0FDQyxhQUFhO0NBQ2IsbUJBQW1CO0NBQ25CLG1CQUFtQjtBQUNwQjs7QUFFQTs7OztDQUlDLGdCQUFnQjtBQUNqQjs7QUFFQTtDQUNDLFdBQVc7QUFDWjs7QUFFQTtDQUNDLFdBQVc7QUFDWjs7QUFFQTs7Ozs7OztDQU9DLFdBQVc7QUFDWjs7QUFFQTs7Ozs7O0NBTUMsV0FBVztBQUNaOztBQUVBOzs7OztDQUtDLGNBQWM7Q0FDZCxvRUFBb0U7Q0FDcEUsaUNBQWlDO0FBQ2xDOztBQUVBOzs7Q0FHQyxXQUFXO0FBQ1o7O0FBRUE7O0NBRUMsY0FBYztBQUNmOztBQUVBOzs7Q0FHQyxXQUFXO0FBQ1o7O0FBRUE7O0NBRUMsaUJBQWlCO0FBQ2xCO0FBQ0E7Q0FDQyxrQkFBa0I7QUFDbkI7O0FBRUE7Q0FDQyxZQUFZO0FBQ2IiLCJzb3VyY2VSb290IjoiIn0= */
  </style>
  <style>
  
  </style>
  <style>
    .markdown-body {
      font-family: -apple-system,Segoe UI,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
      box-sizing: border-box;
      min-width: 200px;
      max-width: 980px;
      margin: 0 auto;
      padding: 45px;
    }

    @media not print {
      .markdown-body {
        padding: 45px;
      }

      @media (max-width: 767px) {
        .markdown-body {
          padding: 15px;
        }
      }
    }

    .hf-container {
      color: #24292e;
      line-height: 1.3;
    }

    .markdown-body .highlight pre,
    .markdown-body pre {
      white-space: pre-wrap;
    }
    .markdown-body table {
      display: table;
    }
    .markdown-body img[data-align="center"] {
      display: block;
      margin: 0 auto;
    }
    .markdown-body img[data-align="right"] {
      display: block;
      margin: 0 0 0 auto;
    }
    .markdown-body li.task-list-item {
      list-style-type: none;
    }
    .markdown-body li > [type=checkbox] {
      margin: 0 0 0 -1.3em;
    }
    .markdown-body input[type="checkbox"] ~ p {
      margin-top: 0;
      display: inline-block;
    }
    .markdown-body ol ol,
    .markdown-body ul ol {
      list-style-type: decimal;
    }
    .markdown-body ol ol ol,
    .markdown-body ol ul ol,
    .markdown-body ul ol ol,
    .markdown-body ul ul ol {
      list-style-type: decimal;
    }
  </style>
  <style>.markdown-body a.footnote-ref {
  text-decoration: none;
}

.footnotes {
  font-size: .85em;
  opacity: .8;
}

.footnotes li[role="doc-endnote"] {
  position: relative;
}

.footnotes .footnote-back {
  position: absolute;
  font-family: initial;
  top: .2em;
  right: 1em;
  text-decoration: none;
}

.inline-math.invalid,
.multiple-math.invalid {
  color: rgb(255, 105, 105);
}

.toc-container {
  width: 100%;
}

.toc-container .toc-title {
  font-weight: 700;
  font-size: 1.2em;
  margin-bottom: 0;
}

.toc-container li,
.toc-container ul,
.toc-container ul li {
  list-style: none !important;
}

.toc-container > ul {
  padding-left: 0;
}

.toc-container ul li span {
  display : flex;
}

.toc-container ul li span a {
  color: inherit;
  text-decoration: none;
}
.toc-container ul li span a:hover {
  color: inherit;
  text-decoration: none;
}

.toc-container ul li span span.dots {
  flex: 1;
  height: 0.65em;
  margin: 0 10px;
  border-bottom: 2px dotted black;
}

/*# sourceURL=webpack://./src/muya/lib/assets/styles/exportStyle.css */
/*# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIndlYnBhY2s6Ly8uL3NyYy9tdXlhL2xpYi9hc3NldHMvc3R5bGVzL2V4cG9ydFN0eWxlLmNzcyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQTtFQUNFLHFCQUFxQjtBQUN2Qjs7QUFFQTtFQUNFLGdCQUFnQjtFQUNoQixXQUFXO0FBQ2I7O0FBRUE7RUFDRSxrQkFBa0I7QUFDcEI7O0FBRUE7RUFDRSxrQkFBa0I7RUFDbEIsb0JBQW9CO0VBQ3BCLFNBQVM7RUFDVCxVQUFVO0VBQ1YscUJBQXFCO0FBQ3ZCOztBQUVBOztFQUVFLHlCQUF5QjtBQUMzQjs7QUFFQTtFQUNFLFdBQVc7QUFDYjs7QUFFQTtFQUNFLGdCQUFnQjtFQUNoQixnQkFBZ0I7RUFDaEIsZ0JBQWdCO0FBQ2xCOztBQUVBOzs7RUFHRSwyQkFBMkI7QUFDN0I7O0FBRUE7RUFDRSxlQUFlO0FBQ2pCOztBQUVBO0VBQ0UsY0FBYztBQUNoQjs7QUFFQTtFQUNFLGNBQWM7RUFDZCxxQkFBcUI7QUFDdkI7QUFDQTtFQUNFLGNBQWM7RUFDZCxxQkFBcUI7QUFDdkI7O0FBRUE7RUFDRSxPQUFPO0VBQ1AsY0FBYztFQUNkLGNBQWM7RUFDZCwrQkFBK0I7QUFDakMiLCJzb3VyY2VSb290IjoiIn0= */</style>
  <style>.markdown-body{}pre.front-matter{display:none!important;}</style>
</head>
<body>
  <article class="markdown-body"><h1 class="atx" id="informe-de-caso">Informe de caso</h1>
<h2 class="atx" id="trabajo-final-del-modulo-de-transcriptomica-202223">Trabajo final del módulo de transcriptómica 2022/23</h2>
<h3 class="atx" id="alumno-miguel-ramon-alonso">Alumno: Miguel Ramón Alonso</h3>
<h2 class="atx" id="introduccion">INTRODUCCIÓN</h2>
<p>El objetivo de este trabajo es que el alumno demuestre los conocimientos obtenidos acerca del análisis de RNA-seq. Para ello deberá <strong>redactar un informe</strong> en el que se expliquen los datos de partida y se extraiga una <strong>conclusión de los resultados</strong>. Así mismo, el alumno debe <strong>detallar el proceso de análisis</strong> indicando el <em>software</em> (incluída la versión) empleado, así como los parámetros utilizados en cada uno de los pasos. En caso de que hubiera que eliminar muestras por motivos técnicos o biológicos, el alumno debe indicar y justificar el por qué en cada caso. Se pueden introducir <em>code chunks</em> e imágenes para apoyar el informe. De manera alternativa, puede aportarse todo el código en forma de repositorio público. Se han planteado <strong>5 preguntas (10 puntos en total)</strong> para guiar la redacción del informe.</p>
<p>El trabajo consta de dos apartados en los que se utilizarán datos de un experimento en el que disponemos de 24 cultivos primarios de tumores paratiroideos negativos para receptores de estrógenos alfa (ERα). Las muestras, procedentes de 4 pacientes diferentes, se han tratado con dos fármacos diferentes: diarilpropionitrilo (DPN) o 4-hidroxitamoxifeno (OHT) a 24h o 48h. El DPN es un agonista del ERα mientras que el OHT es un inhibidor competitivo de los receptores de estrógenos.</p>
<p>El <strong>primer apartado (3 preguntas)</strong> abarca los pasos de control de calidad y de fuentes de contaminación, <em>trimming</em>, alineamiento y cuantificación para obtener cuentas crudas y normalizadas a partir de un <strong>subset de ficheros fastq</strong>. El <strong>segundo apartado (2 preguntas)</strong> parte de la <strong>matriz completa de cuentas crudas</strong> y está enfocado a realizar un control de calidad biológico, detectar los genes diferencialmente expresados entre condiciones y los pathways enriquecidos en cada una de ellas.</p>
<h4 class="atx" id="apartado-1">Apartado 1</h4>
<p>El dataset original consta de 27 muestras paired-end depositadas en SRA. Con el fin de poder abordar las cuestiones planteadas a en el primer apartado sólo es necesario descargar 2 muestras (SRR479052 y SRR479054), es decir cuatro ficheros fastq. Además, en este repositorio se proporciona:</p>
<ul>
<li>Un fichero fasta con la secuencia de la referencia genómica, en este caso correspondiente al cromosoma 21 humano (ensamblaje GRCh38).</li>
<li>Un fichero GTF con la anotación génica para los genes del cromosoma 21 (GRCh38.ensembl.109).</li>
</ul>
<p>Se pide realizar un análisis de dichas muestras similar al realizado en clase, considerando los siguientes puntos:</p>
<p><strong>Pregunta 1 (1.5 puntos):</strong> Realizar control de calidad de dichas muestras con el programa FastQC, incluir plots más reseñables y comentar cada uno de los apartados. De manera complementaria se podrá realizar un analisis de contaminación con el programa FastQScreen.</p>
<p><strong>Pregunta 2 (1.5 puntos):</strong> Para poder llevar a cabo el alineamiento de las muestras en vuestros ordenadores será necesario trabajar con archivos reducidos correspondientes al cromosoma 21. Se requiere el indexado de la secuencia de este cromosoma, así como el alineamiento de las muestras a dicha referencia. Para ello se podrá utilizar el alineador HISAT2 utilizado en clase u otros (alineadores o pseudoalineadores). Comentar cada uno de los comandos y parámetros empleados, justificando su uso.</p>
<p><strong>Pregunta 3 (1.5 puntos):</strong> Una vez generados los archivos alineados se reportarán las estadísticas de alineamiento y se procederá a la cuantificación de la expresión utilizando el archivo GTF correspondiente. Para ello se podrá utilizar HTSeq u otras herramientas de cuantificación. En cualquier caso, detallar y justificar los comandos y parámetros empleados para ello.</p>
<h4 class="atx" id="apartado-2">Apartado 2</h4>
<p>En este repositorio se proporcionan todos los inputs del Apartado 2:</p>
<ul>
<li>La matriz de cuentas crudas para los 24 cultivos analizados.</li>
<li>Data frame con los metadatos asociados al experimento.</li>
<li>GMT para realizar un GSEA.</li>
</ul>
<p><strong>Pregunta 4 (3 puntos):</strong> ¿Qué genes se encuentran diferencialmente expresados entre las muestras pertenecientes al grupo tratado con OHT con respecto al control tras 24h? ¿Y en el caso de las muestras tratadas con DPN, también tras 24h? Como parte de la respuesta a esta pregunta, podéis entregar una o más tablas adjuntas donde se incluyan los genes diferencialmente expresados, indicando el criterio o los criterios que habéis seguido para filtrar los resultados, así como cualquier otro gráfico o gráficos que hayáis elaborado durante el análisis.</p>
<p><strong>Pregunta 5 (2.5 puntos):</strong> Nuestro colaborador ha comparado las muestras tratadas con DPN tras 48h con las muestras control. Nos ha llamado para contarnos que los cambios de expresión tras 48h son mucho más evidentes y nos preguntamos si el DPN produce algún efecto en las primeras 24h. Para contestar esta pregunta, le pedimos que genere un GMT (input) con los genes más expresados en las muestras tratadas tras 48h (DPN_perturbed) y los genes más expresados en la muestra control (DPN_unperturbed). Realiza un análisis con GSEA para determinar el efecto del tratamiento a las 24h. ¿A qué conclusión llegas? Incluid una tabla con los resultados del análisis, destacando las columnas con los valores utilizados para extraer vuestras conclusiones. También incluid los gráficos característicos de este tipo de análisis.</p>
<h2 class="atx" id="metodos">MÉTODOS</h2>
<p>Para dar respuesta  a las diferentes preguntas que propone el ejercicio, se decidió realizar un pipeline interactivo, que permitiese al usuario introducir cualquier muestra SRA, la cual sería descargada, desempaquetada en los correspondientes archivos FASTQ, analizada, preprocesada, alineada y post-procesada utilizando diferentes metodologías.</p>
<h4 class="atx" id="descarga-del-genoma-downloadsh">Descarga del genoma <code>download.sh</code></h4>
<p>Se utilizó la herramienta de sistema <em>wget</em>, integrando un <em>input control</em> que confirme la correcta introducción del formato de SRA por parte del usuario (<strong>if [[ "SRAentry" =~ ^[Ss]if [[ "SRAentry" =~ ^[Ss]RR[0-9]{6}RR[0-9]{6} ]]</strong>), y los parámetros específicos de <em>wget</em> <strong>-nc (no clobber, evita descargas en duplicado)</strong>, <strong>-P (designa el directorio de descarga)</strong>, y <strong>--content-disposition (mantiene la integridad del formato de descarga, evitando el que por error descargue la el enlace como HTML)</strong>.</p>
<pre><code class="fenced-code-block language-shell"> <span class="token keyword">if</span> <span class="token punctuation">[</span><span class="token punctuation">[</span> <span class="token string">"<span class="token variable">$SRAentry</span>"</span> <span class="token operator">=~</span> ^<span class="token punctuation">[</span>Ss<span class="token punctuation">]</span>RR<span class="token punctuation">[</span><span class="token number">0</span>-9<span class="token punctuation">]</span><span class="token punctuation">{</span><span class="token number">6</span><span class="token punctuation">}</span>$ <span class="token punctuation">]</span><span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
                <span class="token builtin class-name">echo</span> <span class="token string">"Matched SRA accession <span class="token variable">$SRAentry</span>"</span>
                <span class="token punctuation">(</span>wget -nc -P <span class="token variable">$sample_dir</span>/SRA --content-disposition https://sra-pub-run-odp.s3.amazonaws.com/sra/<span class="token variable">$SRAentry</span>/<span class="token variable">$SRAentry</span><span class="token punctuation">)</span> <span class="token operator">&amp;</span> spinner
                <span class="token builtin class-name">echo</span> <span class="token string">"Finished downloading <span class="token variable">$SRAentry</span>"</span></code></pre>
<p>Al igual que para el resto de procesos que se realizan a lo largo del pipeline, se añadió un pequeño spinner o reloj para embellecer la espera y para dar al usuario seguridad de que el programa no se ha congelado (ver <code>&amp; spinner</code> y <code>spinner.sh</code>. )</p>
<h4 class="atx" id="desempaquetado-de-ficheros-fastq-downloadsh">Desempaquetado de ficheros FASTQ <code>download.sh</code></h4>
<p>Se utilizó la herramienta fasterq-dump para desempaquetar el fichero SRA en los consecuentes fFASTQ y rFASTQ, introduciendo una comprobación de existencia del fichero. Los parámetros utilizados fueron <strong>-fp</strong> (forzar extracción y mostrar barra de procesamiento).</p>
<pre><code class="fenced-code-block language-shell"><span class="token keyword">if</span> <span class="token punctuation">[</span> <span class="token string">"<span class="token variable"><span class="token variable">$(</span><span class="token function">ls</span> -A $sample_dir/dumped_fastq/$SRAentry<span class="token variable">)</span></span>"</span> <span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
                        <span class="token builtin class-name">echo</span> -e <span class="token string">"Fastq already dumped for <span class="token variable">$SRAentry</span>, skipping.<span title="\n" class="token entity">\n</span>"</span>

                    <span class="token keyword">else</span>
                        <span class="token punctuation">(</span>fasterq-dump -fp -O <span class="token variable">$sample_dir</span>/dumped_fastq/<span class="token variable">$SRAentry</span> <span class="token variable">$SRAentry</span><span class="token punctuation">)</span> <span class="token operator">&amp;</span> spinner <span class="token variable">$!</span></code></pre>
<p>Una vez descargados, se le pregunta al usuario si desea realizar un control de calidad de los FASTQ brutos. En caso afirmativo se procede al paso de QC; en caso contrario, se procede directamente al menú general.</p>
<pre><code class="fenced-code-block language-shell"><span class="token builtin class-name">read</span> -rp <span class="token string">"Would you like to perform a QC analysis (FastQC and FastQScreen)? (Y/n): "</span> performQC

<span class="token keyword">for</span> <span class="token for-or-select variable">SRAentry</span> <span class="token keyword">in</span> <span class="token string">"<span class="token variable">${SRAentries<span class="token punctuation">[</span>@<span class="token punctuation">]</span>}</span>"</span><span class="token punctuation">;</span> <span class="token keyword">do</span>
    <span class="token keyword">for</span> <span class="token for-or-select variable">sid</span> <span class="token keyword">in</span> <span class="token variable"><span class="token variable">$(</span><span class="token function">find</span> $sample_dir/dumped_fastq/$SRAentry -type f -name <span class="token string">'*.fastq'</span><span class="token variable">)</span></span><span class="token punctuation">;</span> <span class="token keyword">do</span>
        <span class="token assign-left variable">base_sid</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token function">basename</span> $sid <span class="token operator">|</span> <span class="token function">cut</span> -d<span class="token string">"_"</span> -f1<span class="token variable">)</span></span>

        <span class="token keyword">if</span> <span class="token punctuation">[</span> <span class="token string">"<span class="token variable"><span class="token variable">$(</span><span class="token function">ls</span> -A $sample_dir/dumped_fastq/$base_sid<span class="token variable">)</span></span>"</span> <span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
            <span class="token keyword">case</span> <span class="token variable">$performQC</span> <span class="token keyword">in</span>
                <span class="token punctuation">[</span>Yy<span class="token punctuation">]</span>* <span class="token punctuation">)</span>
                    <span class="token function">bash</span> scripts/qc.sh <span class="token variable">$sid</span> <span class="token string">"out/qc/fastqc"</span> <span class="token string">"out/qc/fastq_screen"</span>
                <span class="token punctuation">;</span><span class="token punctuation">;</span>
                <span class="token punctuation">[</span>Nn<span class="token punctuation">]</span>* <span class="token punctuation">)</span>
                    <span class="token builtin class-name">echo</span> -e <span class="token string">"<span title="\n" class="token entity">\n</span>Skipping QC analysis...<span title="\n" class="token entity">\n</span>"</span>
                    <span class="token builtin class-name">break</span> 
                <span class="token punctuation">;</span><span class="token punctuation">;</span>
            <span class="token keyword">esac</span>
        <span class="token keyword">fi</span>
    <span class="token keyword">done</span>
<span class="token keyword">done</span>


<span class="token comment"># Results visualization</span>
<span class="token keyword">case</span> <span class="token variable">$performQC</span> <span class="token keyword">in</span>
<span class="token punctuation">[</span>Yy<span class="token punctuation">]</span>* <span class="token punctuation">)</span>
    <span class="token builtin class-name">printf</span> <span class="token string">"<span class="token variable">${GREEN}</span><span title="\n" class="token entity">\n</span>Now, take your time to give a look to the QC analysis.<span title="\n" class="token entity">\n</span><span class="token variable">${NC}</span>"</span>
    <span class="token builtin class-name">echo</span> <span class="token string">"When you are ready, press any key to continue..."</span>
    <span class="token builtin class-name">read</span> -n <span class="token number">1</span> -s -r -p <span class="token string">""</span>
<span class="token punctuation">;</span><span class="token punctuation">;</span>
<span class="token keyword">esac</span></code></pre>
<h4 class="atx" id="opt-1-realizar-control-de-calidad-de-fastq-brutos-qcsh">OPT 1: Realizar control de calidad de FASTQ brutos <code>qc.sh</code></h4>
<p>Para el control de calidad de las muestras, se utilizó la combinación de programas <em>fastQC</em> y <em>fastq-screen</em>. El script tiene dos posibilidades de uso; por una parte, está pensado para ser utilizado dentro de la ejecución del pipeline <code>pipeline.sh</code>, como se ha descrito hasta ahora en la metodología; por otra, y al igual que el resto de herramientas que se describan en adelante, permite un uso independiente con cualquier fichero correspondiente que disponga el usuario, mediante la siguiente porción de código:</p>
<pre><code class="fenced-code-block language-shell"><span class="token comment"># Comprobation for independent use of script</span>
<span class="token keyword">if</span> <span class="token punctuation">[</span> <span class="token string">"<span class="token variable">$#</span>"</span> -ne <span class="token number">3</span> <span class="token punctuation">]</span>
<span class="token keyword">then</span>
    <span class="token builtin class-name">printf</span> <span class="token string">"<span class="token variable">${RED}</span>Usage: <span class="token variable">$1</span> &lt;fastq_file&gt; <span class="token variable">$2</span> &lt;fastqc_out_dir&gt; <span class="token variable">$3</span> &lt;fastqscreen_out_dir&gt; <span class="token variable">${NC}</span><span title="\n" class="token entity">\n</span>"</span>
    <span class="token builtin class-name">exit</span> <span class="token number">1</span>
<span class="token keyword">fi</span> </code></pre>
<p>En este caso, el programa comprueba la inserción de 3 parámetros junto con la ejecución del código (sucede por defecto con el pipeline general), y en caso contrario, avisa al usuario del uso correcto de la herramienta. El ${RED} hace referencia a una serie de variables declaradas en la cabecera de cada script, indicando diferentes colores que van a ser usados para dar mayor comprehensibilidad al código.</p>
<h5 class="atx" id="fastqc">FASTQC</h5>
<p>Realiza las comprobaciones pertinentes y después ejecuta el programa <em>fastQC</em> con los parámetros <strong>-o (directorio de salida del informe generado)</strong> y <strong>2&gt;&amp;1 &gt;/dev/null | tail -n +2 2&gt;&amp;1 log/qc/fastqc.log (redirige la salida de consola a un fichero externo y elimina una línea innecesaria que puede dar lugar a confusión).</strong></p>
<pre><code class="fenced-code-block language-shell">fastqc -o <span class="token variable">$fastqc_dir</span>/<span class="token variable">$cut_sid</span> <span class="token variable">$sid</span> <span class="token operator"><span class="token file-descriptor important">2</span>&gt;</span><span class="token file-descriptor important">&amp;1</span> <span class="token operator">&gt;</span>/dev/null <span class="token operator">|</span> <span class="token function">tail</span> -n +2 <span class="token operator"><span class="token file-descriptor important">2</span>&gt;</span><span class="token file-descriptor important">&amp;1</span> log/qc/fastqc.log</code></pre>
<h5 class="atx" id="fastq-screen">FASTQ-SCREEN</h5>
<p>Realiza las comprobaciones pertinentes</p>
<pre><code class="fenced-code-block language-shell"><span class="token keyword">if</span> <span class="token punctuation">[</span> -f <span class="token variable">$fastqscreen_dir</span>/<span class="token variable">$cut_sid</span>/<span class="token variable">${nofastq_sid}</span>_screen.html <span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
    <span class="token builtin class-name">echo</span> -e <span class="token string">"FastQScreen analysis already performed for <span class="token variable">$nofastq_sid</span>, skipping analysis.<span title="\n" class="token entity">\n</span>"</span> </code></pre>
<p>y otorga al usuario la posibilidad de descargar los índices de referencia utilizados por la herramienta para el uso de la misma, añadiendo un aviso para el usuario sobre el peso de los mismos y el tiempo de descarga que requieren</p>
<pre><code class="fenced-code-block language-shell"><span class="token keyword">else</span>
    <span class="token assign-left variable">screen_gen</span><span class="token operator">=</span><span class="token string">"res/fastq_screen_samples/FastQ_Screen_Genomes"</span>
    <span class="token keyword">if</span> <span class="token punctuation">[</span> <span class="token operator">!</span> <span class="token string">"<span class="token variable"><span class="token variable">$(</span><span class="token function">ls</span> -A $screen_gen<span class="token variable">)</span></span>"</span> <span class="token punctuation">]</span><span class="token punctuation">;</span> <span class="token keyword">then</span>
        <span class="token builtin class-name">echo</span> <span class="token string">"No reference genomes could be found."</span>
        <span class="token builtin class-name">echo</span> -e <span class="token string">"<span title="\n" class="token entity">\n</span>Would you like to download genome indexes from database?"</span>
        <span class="token builtin class-name">echo</span> -e <span class="token string">"It takes long AND it takes space. Note that if you don't download the references,
you have to manually curate fastqscreen.conf file and point the program to it"</span> 
        <span class="token builtin class-name">read</span> -rp <span class="token string">"Continue? (Y/n): "</span> genDownload

        <span class="token keyword">case</span> <span class="token variable">$genDownload</span> <span class="token keyword">in</span>
            <span class="token punctuation">[</span>Yy<span class="token punctuation">]</span>* <span class="token punctuation">)</span>
                <span class="token builtin class-name">echo</span> -e <span class="token string">"<span title="\n" class="token entity">\n</span>Downloading genomes...<span title="\n" class="token entity">\n</span>"</span> 
                fastq_screen --get_genomes --outdir <span class="token string">"res/fastq_screen_samples"</span>
            <span class="token punctuation">;</span><span class="token punctuation">;</span>
            <span class="token punctuation">[</span>Nn<span class="token punctuation">]</span>* <span class="token punctuation">)</span>
                <span class="token builtin class-name">echo</span> -e <span class="token string">"<span title="\n" class="token entity">\n</span>Skipping genome download...<span title="\n" class="token entity">\n</span>"</span>
            <span class="token punctuation">;</span><span class="token punctuation">;</span>
        <span class="token keyword">esac</span></code></pre>
<p>para después correr la herramienta utilizando los parámetros <strong>--conf  (apunta al fichero de configuración)</strong>, <strong>--tag (etiqueta las lecturas en función del indice con el que casen)</strong>, -<strong>-aligner (especifica el alineador a utilizar, el cual debe corresponderse con el usado para la generación de los índices; en el caso de los usados de serie por <em>fastqscreen</em>, bowtie2 )</strong>, <strong>--subset (tamaño de sampleo, para no usar todo el fastq)</strong>,<strong>--threads (numero de cores a utilizar)</strong>, y <strong>--outdir (indica el directorio de salida).</strong></p>
<pre><code class="fenced-code-block language-shell">fastq_screen --conf <span class="token string">"<span class="token variable">$screen_gen</span>/fastq_screen.conf"</span> <span class="token punctuation">\</span>
    --tag --aligner bowtie2 --subset <span class="token number">100000</span> --threads <span class="token number">14</span> <span class="token punctuation">\</span>
    --outdir <span class="token string">"<span class="token variable">$fastqscreen_dir</span>/<span class="token variable">$cut_sid</span>"</span> <span class="token variable">$sid</span> <span class="token operator"><span class="token file-descriptor important">2</span>&gt;</span><span class="token file-descriptor important">&amp;1</span> log/qc/fastqscreen.log<span class="token punctuation">)</span> <span class="token operator">&amp;</span> spinner <span class="token variable">$!</span></code></pre>
<h4 class="atx" id="opt-2-menu-de-opciones-y-eleccion-de-workflow-pipelinesh">OPT 2: Menú de Opciones y elección de workflow <code>pipeline.sh</code></h4>
<p>Tanto como si se ha decidido no realizar un control de calidad de las muestras como si se ha hecho, el siguiente paso consiste en el procesamiento de la muestra a través de numeras opciones de workflow en función de los intereses del investigador, a través de un menú interactivo.</p>
<p>Éste se encuentra dividido en dos categorías; por una parte, workflows sin preprocesamiento para alineamiento rápido en muestras de muy alta calidad; y por otra, diferentes workflows que incluyen pre-procesado previo de las mismas.</p>
<pre><code class="fenced-code-block language-shell"><span class="token comment"># Different pipelines for different workflows</span>
<span class="token builtin class-name">echo</span> -e <span class="token string">"<span class="token variable">${GREEN}</span><span title="\n" class="token entity">\n</span>WORKFLOW PIPELINE MENU<span title="\n" class="token entity">\n</span><span class="token variable">${NC}</span>
<span class="token variable">${RED}</span>Default cores set to 14, change accordingly<span title="\n" class="token entity">\n</span><span class="token variable">${NC}</span>"</span>

<span class="token builtin class-name">echo</span> -e <span class="token string">"
<span class="token variable">${YELLOW}</span>BASIC WORKFLOWS FOR HIGH QUALITY READS, NO PRE-PROCESSING<span title="\n" class="token entity">\n</span><span class="token variable">${NC}</span>
<span title="\t" class="token entity">\t</span>WORKFLOW 1 (Using STAR aligner)<span title="\n" class="token entity">\n</span>
<span title="\t" class="token entity">\t</span>WORKFLOW 2 (Using HISAT2 aligner)<span title="\n" class="token entity">\n</span>
<span title="\t" class="token entity">\t</span>WORKFLOW 3 (Using SALMON pseudo-aligner)<span title="\n" class="token entity">\n</span>
<span title="\t" class="token entity">\t</span>WORKFLOW 4 (Using KALLISTO pseudo-aligner)<span title="\n" class="token entity">\n</span>
<span class="token variable">${YELLOW}</span>ADVANCED WORKFLOWS INCLUDING PREPROCESSING<span title="\n" class="token entity">\n</span><span class="token variable">${NC}</span>
<span title="\t" class="token entity">\t</span>WORKFLOW 5 (Toolset: FastQScreen, Cutadapt, STAR)<span title="\n" class="token entity">\n</span>
<span title="\t" class="token entity">\t</span>WORKFLOW 6 (Toolset: FastQScreen, Cutadapt, HISAT2)<span title="\n" class="token entity">\n</span>
<span title="\t" class="token entity">\t</span>WORKFLOW 7 (Toolset: FastQScreen, Cutadapt, SALMON)<span title="\n" class="token entity">\n</span>
<span title="\t" class="token entity">\t</span>WORKFLOW 8 (Toolset: FastQScreen, Cutadapt, KALLISTO)<span title="\n" class="token entity">\n</span>
<span title="\t" class="token entity">\t</span>WORKFLOW 9 (Toolset: FastQScreen, Fastp, STAR)<span title="\n" class="token entity">\n</span>
<span title="\t" class="token entity">\t</span>WORKFLOW 10 (Toolset: FastQScreen, Fastp, HISAT2)<span title="\n" class="token entity">\n</span>
<span title="\t" class="token entity">\t</span>WORKFLOW 11 (Toolset: FastQScreen, Fastp, SALMON)<span title="\n" class="token entity">\n</span>
<span title="\t" class="token entity">\t</span>WORKFLOW 12 (Toolset: FastQScreen, Fastp, KALLISTO)<span title="\n" class="token entity">\n</span>"</span>

<span class="token builtin class-name">read</span> -rp <span class="token string">"Option: "</span> menuOp</code></pre>
<h5 class="atx" id="herramientas-de-pre-procesamiento">Herramientas de pre-procesamiento</h5>
<h6 class="atx" id="cutadapt">Cutadapt</h6>
<p>Cutadapt es una herramienta que permite el cribado de las lecturas por calidad, así como la remoción de adaptadores usados en la secuenciación. Debido al tipo de experimento realizado, se utilizó con los siguientes parámetros (los cuales deben ser modificados acordes al experimento a analizar):</p>
<ul>
<li><strong>-a, -A, -g y -G</strong> representan respectivamente adaptadores hallados en los extremos 5' y 3' de las muestras forward (minúscula) y reverse (mayúscula). Se eliminaron los diferentes adaptadores encontrados en el análisis FasQC + típicos en secuenciación Illumina PE (enlaces), así como residuos PolyA y PolyG.</li>
<li><strong>-q</strong> es el umbral de calidad que debe superar una lectura para ser considerada válida. Se fijó en 15 por concordancia con el umbral por defecto de FastP.</li>
<li><strong>-o</strong>  y <strong>-p</strong> son las rutas de salida del fichero procesado, seguidas en este caso de los ficheros a procesar (f_path y r_path).</li>
<li><strong>--cores</strong> fija el número de nucleos</li>
<li><strong>-m</strong> fija la longitud mínima que debe tener una lectura para ser considerada válida. Se fijó en 15 por concordancia con el umbral por defecto de FastP.</li>
<li><strong>--discard-untrimmed</strong> toma como válidas solo aquellas lecturas que hayan sido preprocesadas.</li>
</ul>
<pre><code class="fenced-code-block language-shell">cutadapt -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT <span class="token punctuation">\</span>
            -a CTGTCTCTTATACACATCT -A CTGTCTCTTATACACATCT <span class="token punctuation">\</span>
            -g TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG -G GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG <span class="token punctuation">\</span>
            -a CTGTCTCTTATACACATCT<span class="token punctuation">..</span>.AGATGTGTATAAGAGACAG -a TGGAATTCTCGGGTGCCAAGG <span class="token punctuation">\</span>
            -a <span class="token string">"G{10}"</span> -A <span class="token string">"G{10}"</span> -g <span class="token string">"G{10}"</span> -G <span class="token string">"G{10}"</span> -g <span class="token string">"A{10}"</span> -G <span class="token string">"A{10}"</span> -q <span class="token number">15</span> <span class="token punctuation">\</span>
            -o <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">${f_sid}</span>_trimmed.fastq"</span> -p <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">${r_sid}</span>_trimmed.fastq"</span> <span class="token punctuation">\</span>
            <span class="token string">"<span class="token variable">$f_path</span>"</span> <span class="token string">"<span class="token variable">$r_path</span>"</span> --cores <span class="token number">14</span> -m <span class="token number">15</span> --discard-untrimmed <span class="token operator">&gt;</span> <span class="token string">"<span class="token variable">$logdir</span>"</span>/log.txt</code></pre>
<p>Tras esto, se le da la posibilidad al usuario de realizar un nuevo control de calidad de las muestras procesadas.</p>
<pre><code class="fenced-code-block language-shell"><span class="token comment"># Re-run quality control</span>
    <span class="token builtin class-name">read</span> -rp <span class="token string">"Would you like to re-run QC report for your trimmed samples? (Y/n) "</span> runqc

    <span class="token keyword">for</span> <span class="token for-or-select variable">trimmed_sid</span> <span class="token keyword">in</span> <span class="token variable"><span class="token variable">$(</span><span class="token function">find</span> $outdir -type f -name <span class="token string">"*_trimmed.fastq"</span><span class="token variable">)</span></span><span class="token punctuation">;</span><span class="token keyword">do</span>
        <span class="token keyword">case</span> <span class="token variable">$runqc</span> <span class="token keyword">in</span>
            <span class="token punctuation">[</span>Yy<span class="token punctuation">]</span>* <span class="token punctuation">)</span>
                <span class="token function">bash</span> scripts/qc.sh <span class="token variable">$trimmed_sid</span> <span class="token string">"out/qc/fastqc"</span> <span class="token string">"out/qc/fastq_screen"</span>
            <span class="token punctuation">;</span><span class="token punctuation">;</span>
        <span class="token keyword">esac</span>
    <span class="token keyword">done</span>

    <span class="token keyword">case</span> <span class="token variable">$runqc</span> <span class="token keyword">in</span>
    <span class="token punctuation">[</span>Yy<span class="token punctuation">]</span>* <span class="token punctuation">)</span>
        <span class="token builtin class-name">printf</span> <span class="token string">"<span class="token variable">${GREEN}</span><span title="\n" class="token entity">\n</span>Now, take your time to give a look to the QC analysis.<span title="\n" class="token entity">\n</span><span class="token variable">${NC}</span>"</span>
        <span class="token builtin class-name">echo</span> <span class="token string">"When you are ready, press any key to continue..."</span>
        <span class="token builtin class-name">read</span> -n <span class="token number">1</span> -s -r -p <span class="token string">""</span>
    <span class="token punctuation">;</span><span class="token punctuation">;</span>
    <span class="token keyword">esac</span></code></pre>
<h6 class="atx" id="fastp">FastP</h6>
<p>Herramienta de control de calidad y preprocesado muy completa, que permite automatizar la eliminación de adaptadores y cribado de las lecturas sin necesidad de hardcodear las secuencias. Se usó con los siguientes parámetros:</p>
<ul>
<li><strong>-i, -o, -I</strong> y <strong>-O</strong> representan los ficheros de entrada (i) y salida (o), para las lecturas forward (minúscula) y reverse (mayúscula).</li>
<li><strong>--detect_adapter_for_pe</strong> detecta automáticamente adaptadores usados en PE.</li>
<li><strong>5</strong> y <strong>3</strong> realizan los cortes en ambos extremos de la lectura, en función de si las bases en estas superan un umbral de calidad determinado.</li>
<li><strong>-D</strong> realiza deduplicación de las muestras.</li>
<li><strong>--dup_calc_accuracy</strong> fijado en 3 como calidad por defecto en deduplicación.</li>
<li><strong>-c</strong> corrige errores de solapamiento en muestras PE.</li>
<li><strong>-p</strong> realiza un análisis de secuencias sobrerrepresentadas.</li>
<li><strong>-h</strong> exporta los resultados en formato HTML para mejor visualización que en JSON por defecto.</li>
<li><strong>-w</strong> fija el número de cores a utilizar.</li>
</ul>
<pre><code class="fenced-code-block language-shell">fastp -i <span class="token variable">$f_path</span> -o <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">${f_sid}</span>_trimmed.fastq"</span> -I <span class="token variable">$r_path</span> -O <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">${r_sid}</span>_trimmed.fastq"</span> <span class="token punctuation">\</span>
 --detect_adapter_for_pe -5 -3 -D --dup_calc_accuracy <span class="token number">3</span> -c -p -P <span class="token number">20</span> <span class="token punctuation">\</span>
 -h <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">${base_sid}</span>.html"</span> -w <span class="token number">14</span> <span class="token operator">&gt;</span> <span class="token string">"<span class="token variable">$logdir</span>"</span>/log.txt</code></pre>
<h5 class="atx" id="herramientas-de-indexado-y-alineamiento">Herramientas de indexado y alineamiento</h5>
<p>Para ver una descripción detallada de las diferencias entre las distintas herramientas de alineamiento, echar un vistazo al manual incluído con las mismas.</p>
<p>Uso independeinte del script de indexado</p>
<pre><code class="fenced-code-block language-shell"><span class="token comment"># Comprobation for independent use of script</span>
<span class="token keyword">if</span> <span class="token punctuation">[</span> <span class="token string">"<span class="token variable">$#</span>"</span> -ne <span class="token number">1</span> <span class="token punctuation">]</span>
<span class="token keyword">then</span>
    <span class="token builtin class-name">printf</span> <span class="token string">"<span class="token variable">${RED}</span>Usage: <span class="token variable">$1</span> &lt;tool&gt; <span class="token variable">${NC}</span><span title="\n" class="token entity">\n</span>"</span>
    <span class="token builtin class-name">echo</span> -e <span class="token string">'tool: "STAR", "HISAT2", "SALMON", "KALLISTO"\n'</span>
    <span class="token builtin class-name">exit</span> <span class="token number">1</span>
<span class="token keyword">fi</span></code></pre>
<p>y de alineamiento:</p>
<pre><code class="fenced-code-block language-shell"><span class="token comment"># Comprobation for independent use of script</span>
<span class="token keyword">if</span> <span class="token punctuation">[</span> <span class="token string">"<span class="token variable">$#</span>"</span> -ne <span class="token number">6</span> <span class="token punctuation">]</span>
<span class="token keyword">then</span>
    <span class="token builtin class-name">printf</span> <span class="token string">"<span class="token variable">${RED}</span>Usage: <span class="token variable">$1</span> &lt;tool&gt; <span class="token variable">$2</span> &lt;f_path&gt; <span class="token variable">$3</span> &lt;r_path&gt; <span class="token variable">$4</span> &lt;outdir&gt; <span class="token variable">$5</span> &lt;logdir&gt; <span class="token variable">$6</span> &lt;workflow&gt;<span class="token variable">${NC}</span><span title="\n" class="token entity">\n</span>"</span>
    <span class="token builtin class-name">echo</span> -e <span class="token string">'tool: "STAR", "HISAT2", "SALMON", "KALLISTO"
workflow: "untrimmed", "cutadapt", "fastp"\n'</span>
    <span class="token builtin class-name">exit</span> <span class="token number">1</span>
<span class="token keyword">fi</span></code></pre>
<h6 class="atx" id="star">STAR</h6>
<p>Parámetros utilizados para el indexado de las muestras:</p>
<ul>
<li>El único parámetro a remarcar es <strong>-genomeSAindex</strong>, el cual fue fijado a partir de las recomendaciones dadas por la propia herramienta en su primera ejecución. Modificar acorde a las condiciones del experimento.</li>
</ul>
<pre><code class="fenced-code-block language-shell">STAR     --runThreadN <span class="token number">14</span> --runMode genomeGenerate --genomeDir <span class="token variable">$outdir</span> <span class="token punctuation">\</span>
                --genomeFastaFiles <span class="token variable">$ref_gen</span> --runRNGseed <span class="token number">1998</span> --genomeSAindexNbases <span class="token number">11</span> <span class="token operator">&gt;</span> <span class="token variable">$logdir</span>/<span class="token variable">$tool</span>.log</code></pre>
<p>Parámetros utilizados para el alineamiento:</p>
<p>A remarcar el parámetro <strong>--outSAMtype</strong> fijado en sorted BAM que nos permite ahorrarnos pasos en la conversión del SAM al BAM. Ejecución posterior de <em>samtools</em> para generar index del fichero de alineamiento y diferentes estadísticas del mismo (generadas con <em>samtools</em> y <em>bamCoverage</em>).</p>
<pre><code class="fenced-code-block language-shell">STAR --runThreadN <span class="token number">14</span> --readFilesIn <span class="token string">"<span class="token variable">$f_path</span>"</span> <span class="token variable">$r_path</span>  <span class="token punctuation">\</span>
            --genomeDir <span class="token string">"<span class="token variable">$index_dir</span>"</span> --outReadsUnmapped Fastx  <span class="token punctuation">\</span>
            --outFileNamePrefix <span class="token string">"<span class="token variable">$outdir</span>/"</span> <span class="token punctuation">\</span>
            --outSAMtype BAM SortedByCoordinate

<span class="token comment"># add more params for statistics (there's a few problems here to solve)</span>
<span class="token assign-left variable">bam_file</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token function">find</span> <span class="token string">"<span class="token variable">$outdir</span>"</span> -type f -name <span class="token string">"*.bam"</span><span class="token variable">)</span></span>
<span class="token punctuation">(</span>samtools stats <span class="token string">"<span class="token variable">$bam_file</span>"</span> <span class="token operator">&gt;</span> <span class="token string">"<span class="token variable">$bam_file</span>"</span>.txt<span class="token punctuation">)</span> <span class="token operator">&amp;</span> spinner <span class="token variable">$!</span>
<span class="token punctuation">(</span>samtools index <span class="token string">"<span class="token variable">$bam_file</span>"</span><span class="token punctuation">)</span> <span class="token operator">&amp;</span> spinner <span class="token variable">$!</span>
<span class="token punctuation">(</span>bamCoverage -b <span class="token string">"<span class="token variable">$bam_file</span>"</span> -o <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.bw"</span> --normalizeUsing BPM<span class="token punctuation">)</span> <span class="token operator">&amp;</span> spinner <span class="token variable">$!</span></code></pre>
<h6 class="atx" id="hisat2">HISAT2</h6>
<p>Ningún parámetro a recalcar en la generación del índice:</p>
<pre><code class="fenced-code-block language-shell">hisat2-build -p <span class="token number">14</span> --seed <span class="token number">1998</span> <span class="token variable">$ref_gen</span> <span class="token variable">$outdir</span>/HISAT2</code></pre>
<p>El alineamiento de la muestra requiere de un mayor número de pasos al utilizar <em>hisat2</em> en su conversión a BAM.</p>
<pre><code class="fenced-code-block language-shell"><span class="token punctuation">(</span>hisat2 --new-summary --summary-file <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.hisat2.summary"</span> <span class="token punctuation">\</span>
    -p <span class="token number">14</span> -x <span class="token string">"<span class="token variable">$index_dir</span>/HISAT2"</span> -1 <span class="token string">"<span class="token variable">$f_path</span>"</span> -2 <span class="token string">"<span class="token variable">$r_path</span>"</span> -k <span class="token number">1</span> -S <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.sam"</span><span class="token punctuation">)</span> <span class="token operator">&amp;</span> spinner <span class="token variable">$!</span>

<span class="token comment"># problema con el pipe</span>
<span class="token comment"># add more params for statistics</span>
<span class="token punctuation">(</span>samtools view -bS <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.sam"</span> <span class="token operator">&gt;</span> <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.bam"</span><span class="token punctuation">)</span> <span class="token operator">&amp;</span> spinner <span class="token variable">$!</span>
<span class="token punctuation">(</span>samtools stats <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.bam"</span> <span class="token operator">&gt;</span> <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.txt"</span><span class="token punctuation">)</span> <span class="token operator">&amp;</span> spinner <span class="token variable">$!</span>
<span class="token punctuation">(</span>samtools <span class="token function">sort</span> --write-index <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.bam"</span> -o <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.sorted.bam"</span><span class="token punctuation">)</span> <span class="token operator">&amp;</span> spinner <span class="token variable">$!</span>
<span class="token punctuation">(</span>samtools stats <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.sorted.bam"</span> <span class="token operator">&gt;</span> <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.sorted.txt"</span><span class="token punctuation">)</span> <span class="token operator">&amp;</span> spinner <span class="token variable">$!</span>
<span class="token punctuation">(</span>bamCoverage -b <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.sorted.bam"</span> -o <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.bw"</span> --normalizeUsing BPM<span class="token punctuation">)</span> <span class="token operator">&amp;</span> spinner <span class="token variable">$!</span></code></pre>
<h6 class="atx" id="salmon">SALMON</h6>
<p>Tanto <em>SALMON</em> como <em>KALLISTO</em> son pseudoalineadores que trabajan de una manera diferente a los dos previamente descritos.</p>
<p>Generación del índice (utilizando exoma de referencia en vez de genoma de referencia):</p>
<pre><code class="fenced-code-block language-shell">salmon index -t <span class="token variable">$ref_cdna</span> -i <span class="token variable">$outdir</span> --gencode -p <span class="token number">14</span> <span class="token operator">&gt;</span> <span class="token variable">$logdir</span>/<span class="token variable">$tool</span>.log</code></pre>
<p>Alineamiento de las muestras:</p>
<ul>
<li><strong>-l</strong> es el tipo de stranding de la muestra. Fijado en A para determinación automática.</li>
<li><strong>-1 y -2</strong> son los ficheros de entrada.</li>
<li><strong>-g</strong> da la opción de añadir fichero de anotación.</li>
<li><strong>--writeMappings</strong> escribe los alineamientos en formato SAM.</li>
</ul>
<pre><code class="fenced-code-block language-shell">salmon quant -i <span class="token string">"<span class="token variable">$index_dir</span>"</span> -l A -1 <span class="token string">"<span class="token variable">$f_path</span>"</span> -2 <span class="token string">"<span class="token variable">$r_path</span>"</span> --validateMappings <span class="token punctuation">\</span>
    -o <span class="token string">"<span class="token variable">$outdir</span>"</span> -g <span class="token string">"<span class="token variable">$ref_gtf</span>"</span> -p <span class="token number">14</span> --writeMappings </code></pre>
<h6 class="atx" id="kallisto">KALLISTO</h6>
<p>Generación de índice:</p>
<pre><code class="fenced-code-block language-shell">kallisto index -i <span class="token variable">$outdir</span>/<span class="token variable">${base_ref_cdna}</span>.idx <span class="token variable">$ref_cdna</span> <span class="token operator">&gt;</span> <span class="token variable">$logdir</span>/<span class="token variable">$tool</span>.log</code></pre>
<p>Pseudoalineamiento:</p>
<ul>
<li><strong>--bias</strong> trata de corrigir errores de sesgo.</li>
<li><strong>--fusion</strong> busca fusiones en Pizzly.</li>
<li><strong>--pseudobam</strong> escribe los pseudoalinamientos con el transcriptoma  a un fichero con formato BAM.</li>
<li><strong>--genomebam</strong> projecta los pseudoalineamientos frente al exoma de referencia.</li>
<li><strong>--gtf</strong> permite la inserción de anotaciones en formato GTF.</li>
</ul>
<pre><code class="fenced-code-block language-shell">kallisto quant -i <span class="token string">"<span class="token variable">$index_dir</span>/<span class="token variable">$base_ref_cdna</span>.idx"</span> --bias --fusion <span class="token punctuation">\</span>
    <span class="token string">"<span class="token variable">$f_path</span>"</span> <span class="token string">"<span class="token variable">$r_path</span>"</span> -o <span class="token string">"<span class="token variable">$outdir</span>"</span> --pseudobam --genomebam <span class="token punctuation">\</span>
    --gtf <span class="token string">"<span class="token variable">$ref_gtf</span>"</span> -t <span class="token number">14</span>

    <span class="token assign-left variable">pseudobam</span><span class="token operator">=</span><span class="token string">"<span class="token variable">$outdir</span>/pseudoalignments.bam"</span>
    bamCoverage -b <span class="token string">"<span class="token variable">$pseudobam</span>"</span> -o <span class="token string">"<span class="token variable">$outdir</span>/<span class="token variable">$base_sid</span>.pseudoalignments.bw"</span> <span class="token punctuation">\</span>
    --normalizeUsing BPM <span class="token operator"><span class="token file-descriptor important">2</span>&gt;</span><span class="token file-descriptor important">&amp;1</span></code></pre>
<h4 class="atx" id="post-procesado-y-conteo-de-muestras-read-summarization-programs">Post-procesado y conteo de muestras (read summarization programs)</h4>
<p>Uso independiente del script:</p>
<pre><code class="fenced-code-block language-shell"><span class="token keyword">if</span> <span class="token punctuation">[</span> <span class="token string">"<span class="token variable">$#</span>"</span> -ne <span class="token number">3</span> <span class="token punctuation">]</span>
<span class="token keyword">then</span>
    <span class="token builtin class-name">printf</span> <span class="token string">"<span class="token variable">${RED}</span>Usage: <span class="token variable">$1</span> &lt;tool&gt; <span class="token variable">$2</span> &lt;bam_file&gt; <span class="token variable">$3</span> &lt;counts_out_dir&gt;<span class="token variable">${NC}</span><span title="\n" class="token entity">\n</span>"</span>
    <span class="token builtin class-name">echo</span> -e <span class="token string">'tool: "featurecounts", "htseq"\n'</span>
    <span class="token builtin class-name">exit</span> <span class="token number">1</span>
<span class="token keyword">fi</span></code></pre>
<h6 class="atx" id="featurecounts">FEATURECOUNTS</h6>
<pre><code class="fenced-code-block language-shell">featureCounts -T <span class="token number">14</span> -s <span class="token number">2</span> -p --countReadPairs -a <span class="token string">"<span class="token variable">$ref_gtf</span>"</span> -t exon -g gene_id <span class="token punctuation">\</span>
    -o <span class="token string">"<span class="token variable">$countdir</span>/counts.txt"</span> <span class="token string">"<span class="token variable">$bam</span>"</span></code></pre>
<h6 class="atx" id="htseq">HTSEQ</h6>
<pre><code class="fenced-code-block language-shell">htseq-count --format<span class="token operator">=</span>bam --mode<span class="token operator">=</span>intersection-nonempty --minaqual<span class="token operator">=</span><span class="token number">10</span> --type<span class="token operator">=</span>exon --idattr<span class="token operator">=</span>gene_id    <span class="token punctuation">\</span>
    --additional-attr<span class="token operator">=</span>gene_name <span class="token string">"<span class="token variable">$bam</span>"</span> <span class="token string">"<span class="token variable">$ref_gtf</span>"</span> <span class="token operator">&gt;</span> <span class="token string">"<span class="token variable">$countdir</span>/<span class="token variable">$base_bam</span>.htseq"</span></code></pre>
<h2 class="atx" id=""></h2>
<h2 class="atx" id="analisis-de-resultados">ANÁLISIS DE RESULTADOS</h2>
<h4 class="atx" id="1er-control-de-calidad-con-fastqc-y-fastqscreen">1ER CONTROL DE CALIDAD CON FASTQC Y FASTQSCREEN</h4>
<ul>
<li><p>Muestra 1_1 y 1_2 (SRR479052)</p>
<p>Diferencias mínimas entre ambas.</p>
<p><strong>FASTQC</strong></p>
</li>
</ul>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-17-59-37-image.png">Clara deficiencia de calidad en las lecturas debido a una alta degradación de rendimiento al final de la secuenciación.</p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-13-10-14-image.png"></p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-13-10-33-image.png"></p>
<p>Presencia de lecturas duplicadas y adaptadores que deben ser eliminados.</p>
<p><strong>FASTQSCREEN</strong></p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-13-28-37-SRR479052_1_screen.png"></p>
<p>Presencia de diferentes genomas contaminantes, vectores, mDNA, adaptadores, etc.</p>
<ul>
<li><p>Muestras 2_1 y 2_2 (SRR479054)</p>
<p>Misma consideración respecto a las diferencias entre ambas</p>
<p><strong>FASTQC</strong></p>
</li>
</ul>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-18-00-07-image.png">Clara deficiencia de calidad en las lecturas debido a una alta degradación de rendimiento al final de la secuenciación.</p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-13-40-21-image.png"></p>
<p>Presencia de lecturas duplicadas y adaptadores que deben ser eliminados.</p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-13-40-30-image.png"></p>
<p>Presencia de lecturas duplicadas y adaptadores que deben ser eliminados.</p>
<p><strong>FASTQSCREEN</strong></p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-13-41-38-SRR479054_1_screen.png"></p>
<p>Presencia de diferentes genomas contaminantes, vectores, mDNA, adaptadores, etc.</p>
<h4 class="atx" id="estadisticas-de-alineamiento-sin-preprocesado">ESTADISTICAS DE ALINEAMIENTO SIN PREPROCESADO</h4>
<p>Se procede a presentar los resultados de un alineador y un pseudoalineador por comparación.</p>
<h5 class="atx" id="hisat2-1">HISAT2</h5>
<p><strong>Muestra SRR479052</strong></p>
<p><u>HISAT2 summary stats:</u></p>
<pre><code class="indented-code-block">Total pairs: 12647534
    Aligned concordantly or discordantly 0 time: 12431709 (98.29%)
    Aligned concordantly 1 time: 211024 (1.67%)
    Aligned concordantly &gt;1 times: 0 (0.00%)
    Aligned discordantly 1 time: 4801 (0.04%)
Total unpaired reads: 24863418
    Aligned 0 time: 24817231 (99.81%)
    Aligned 1 time: 46187 (0.19%)
    Aligned &gt;1 times: 0 (0.00%)
Overall alignment rate: 1.89%</code></pre>
<p><em>Ver gráficas de *plot-bamstats</em> para más información.*</p>
<p><u><em>FeatureCounts summary stats:</em></u></p>
<p>Status    out/aligned/hisat2/SRR479052/20230415141956/SRR479052.sorted.bam
Assigned    62629
Unassigned_Unmapped    12388267
Unassigned_Read_Type    0
Unassigned_Singleton    0
Unassigned_MappingQuality    0
Unassigned_Chimera    0
Unassigned_FragmentLength    0
Unassigned_Duplicate    0
Unassigned_MultiMapping    0
Unassigned_Secondary    0
Unassigned_NonSplit    0
Unassigned_NoFeatures    188257
Unassigned_Overlapping_Length    0
Unassigned_Ambiguity    8381</p>
<p><u><em>HTSEQ stats:</em></u></p>
<p>Warning: 433876 reads with missing mate encountered.
12872221 alignment record pairs processed.</p>
<p><strong>Muestra SRR479054</strong></p>
<p><u><em>HISAT2 summary stats:</em></u>
    Total pairs: 8096699
        Aligned concordantly or discordantly 0 time: 7980123 (98.56%)
        Aligned concordantly 1 time: 114532 (1.41%)
        Aligned concordantly &gt;1 times: 0 (0.00%)
        Aligned discordantly 1 time: 2044 (0.03%)
    Total unpaired reads: 15960246
        Aligned 0 time: 15934432 (99.84%)
        Aligned 1 time: 25814 (0.16%)
        Aligned &gt;1 times: 0 (0.00%)
    Overall alignment rate: 1.60%</p>
<p><em>Ver gráficas de *plot-bamstats</em> para más información.*</p>
<p><u><em>FeatureCounts summary stats:</em></u></p>
<p>Status    out/aligned/hisat2/SRR479054/20230415141956/SRR479054.sorted.bam
Assigned    35982
Unassigned_Unmapped    7955760
Unassigned_Read_Type    0
Unassigned_Singleton    0
Unassigned_MappingQuality    0
Unassigned_Chimera    0
Unassigned_FragmentLength    0
Unassigned_Duplicate    0
Unassigned_MultiMapping    0
Unassigned_Secondary    0
Unassigned_NonSplit    0
Unassigned_NoFeatures    100736
Unassigned_Overlapping_Length    0
Unassigned_Ambiguity    4221</p>
<p><u><em>HTSEQ stats:</em></u></p>
<p>Warning: 23075 reads with missing mate encountered.
987045 alignment record pairs processed.</p>
<h5 class="atx" id="salmon-1">SALMON</h5>
<p><strong>Muestra SRR479052</strong></p>
<p>{
    "read_files": "[ res/samples/dumped_fastq/SRR479052/SRR479052_1.fastq, res/samples/dumped_fastq/SRR479052/SRR479052_2.fastq]",
    "expected_format": "IU",
    "compatible_fragment_ratio": 1.0,
    "num_compatible_fragments": 10656220,
    "num_assigned_fragments": 10656220,
    "num_frags_with_concordant_consistent_mappings": 8995116,
    "num_frags_with_inconsistent_or_orphan_mappings": 1789907,
    "strand_mapping_bias": 0.4986155820558623,
    "MSF": 0,
    "OSF": 0,
    "ISF": 4485105,
    "MSR": 0,
    "OSR": 0,
    "ISR": 4510011,
    "SF": 930565,
    "SR": 859342,
    "MU": 0,
    "OU": 0,
    "IU": 0,
    "U": 0
}</p>
<p><strong>Muestra SRR479054</strong></p>
<p>{
    "read_files": "[ res/samples/dumped_fastq/SRR479054/SRR479054_1.fastq, res/samples/dumped_fastq/SRR479054/SRR479054_2.fastq]",
    "expected_format": "IU",
    "compatible_fragment_ratio": 1.0,
    "num_compatible_fragments": 5986505,
    "num_assigned_fragments": 5986505,
    "num_frags_with_concordant_consistent_mappings": 5204445,
    "num_frags_with_inconsistent_or_orphan_mappings": 850373,
    "strand_mapping_bias": 0.4990628203391524,
    "MSF": 0,
    "OSF": 0,
    "ISF": 2597345,
    "MSR": 0,
    "OSR": 0,
    "ISR": 2607100,
    "SF": 434108,
    "SR": 416265,
    "MU": 0,
    "OU": 0,
    "IU": 0,
    "U": 0
}</p>
<h4 class="atx" id="estadisticas-de-alineamiento-con-preprocesado">ESTADISTICAS DE ALINEAMIENTO CON PREPROCESADO</h4>
<h5 class="atx" id="cutadapt-1">Cutadapt</h5>
<p><strong>Muestra SRR479052</strong></p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-17-59-05-image.png"></p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-17-58-32-image.png"></p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-18-01-02-SRR479052_1_trimmed_screen.png"></p>
<p>Corrección de deficit de calidad por cribado, y eliminación de porciones sobrerrepresentadas y adaptadores.</p>
<p><strong>Muestra SRR479054</strong></p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-18-02-01-image.png"></p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-18-02-31-image.png"></p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-18-02-47-image.png"></p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-18-01-27-SRR479054_1_trimmed_screen.png"></p>
<p>Corrección de deficit de calidad por cribado, y eliminación de porciones sobrerrepresentadas y adaptadores. Presencia de una secuencia sobrerrepresentada desconocida. Tras BLASTear esta secuencia, se asume como contaminación mitocondrial que debería ser eliminada.</p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-18-04-37-image.png"></p>
<h6 class="atx" id="hisat2-2">HISAT2</h6>
<p><strong>Muestra SRR479052</strong></p>
<p>*<u>HISAT2 summary stats:</u>*    </p>
<p>Total pairs: 2180086
        Aligned concordantly or discordantly 0 time: 2132036 (97.80%)
        Aligned concordantly 1 time: 47424 (2.18%)
        Aligned concordantly &gt;1 times: 0 (0.00%)
        Aligned discordantly 1 time: 626 (0.03%)
    Total unpaired reads: 4264072
        Aligned 0 time: 4257451 (99.84%)
        Aligned 1 time: 6621 (0.16%)
        Aligned &gt;1 times: 0 (0.00%)
    Overall alignment rate: 2.36%</p>
<p><u><em>FeatureCounts summary stats:</em></u></p>
<p>Status    out/aligned/hisat2_cutadapt/SRR479052/20230415170851/SRR479052.sorted.bam
Assigned    11710
Unassigned_Unmapped    2125755
Unassigned_Read_Type    0
Unassigned_Singleton    0
Unassigned_MappingQuality    0
Unassigned_Chimera    0
Unassigned_FragmentLength    0
Unassigned_Duplicate    0
Unassigned_MultiMapping    0
Unassigned_Secondary    0
Unassigned_NonSplit    0
Unassigned_NoFeatures    40780
Unassigned_Overlapping_Length    0
Unassigned_Ambiguity    1841</p>
<p><strong>Muestra SRR479054</strong></p>
<p>*<u>HISAT2 summary stats:</u>*    </p>
<p>Total pairs: 975201        Aligned concordantly or discordantly 0 time: 956617 (98.09%)
        Aligned concordantly 1 time: 18333 (1.88%)
        Aligned concordantly &gt;1 times: 0 (0.00%)
        Aligned discordantly 1 time: 251 (0.03%)
    Total unpaired reads: 1913234
        Aligned 0 time: 1909733 (99.82%)
        Aligned 1 time: 3501 (0.18%)
        Aligned &gt;1 times: 0 (0.00%)
    Overall alignment rate: 2.09%</p>
<p><u><em>FeatureCounts summary stats:</em></u></p>
<p>Status    /opt/temp_bioinfo/transcriptomic-final-exercise/out/aligned/hisat2_cutadapt/SRR479054/20230415170851/SRR479054.sorted.bam
Assigned    4841
Unassigned_Unmapped    953267
Unassigned_Read_Type    0
Unassigned_Singleton    0
Unassigned_MappingQuality    0
Unassigned_Chimera    0
Unassigned_FragmentLength    0
Unassigned_Duplicate    0
Unassigned_MultiMapping    0
Unassigned_Secondary    0
Unassigned_NonSplit    0
Unassigned_NoFeatures    16417
Unassigned_Overlapping_Length    0
Unassigned_Ambiguity    676</p>
<h6 class="atx" id="salmon-2">SALMON</h6>
<p><strong>Muestra SRR479052</strong></p>
<p>{
    "read_files": "[ out/trimmed/cutadapt/SRR479052/SRR479052_1_trimmed.fastq, out/trimmed/cutadapt/SRR479052/SRR479052_2_trimmed.fastq]",
    "expected_format": "IU",
    "compatible_fragment_ratio": 1.0,
    "num_compatible_fragments": 1973597,
    "num_assigned_fragments": 1973597,
    "num_frags_with_concordant_consistent_mappings": 1913317,
    "num_frags_with_inconsistent_or_orphan_mappings": 81890,
    "strand_mapping_bias": 0.49651782741699365,
    "MSF": 0,
    "OSF": 0,
    "ISF": 949996,
    "MSR": 0,
    "OSR": 0,
    "ISR": 963321,
    "SF": 32170,
    "SR": 49720,
    "MU": 0,
    "OU": 0,
    "IU": 0,
    "U": 0
}</p>
<p><strong>Muestra SRR479054</strong></p>
<p>{
    "read_files": "[ out/trimmed/cutadapt/SRR479054/SRR479054_1_trimmed.fastq, out/trimmed/cutadapt/SRR479054/SRR479054_2_trimmed.fastq]",
    "expected_format": "IU",
    "compatible_fragment_ratio": 1.0,
    "num_compatible_fragments": 840148,
    "num_assigned_fragments": 840148,
    "num_frags_with_concordant_consistent_mappings": 805230,
    "num_frags_with_inconsistent_or_orphan_mappings": 43832,
    "strand_mapping_bias": 0.49611912124486176,
    "MSF": 0,
    "OSF": 0,
    "ISF": 399490,
    "MSR": 0,
    "OSR": 0,
    "ISR": 405740,
    "SF": 16826,
    "SR": 27006,
    "MU": 0,
    "OU": 0,
    "IU": 0,
    "U": 0
}</p>
<h5 class="atx" id="fastp-1">FASTP</h5>
<p>Ambos informes se incluyen junto con este para su visualización, ya que contienen mucha más información que los de FASTQ.</p>
<p><strong>Muestra SRR479052</strong></p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-18-27-10-image.png"></p>
<p><strong>Muestra SRR479054</strong></p>
<p><img alt="" src="file:///home/miguel/.var/app/com.github.marktext.marktext/config/marktext/images/2023-04-15-18-18-35-image.png"></p>
<h5 class="atx" id="-1"></h5>
<p>Los alineamientos e indices se generan de la misma manera que la anterior. A pesar de que los resultados tras procesar las lecturas son distintos, se decide obviarlos de cara al informe ya que se sale de lo que se está pidiendo. </p>
</article>
</body>
</html>