---
title: "Differential analysis"
author: "Miguel Ramón Alonso"
date: "2023-04-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# DeSEQ2

This is an R Markdown document which contains a fully working script for differential analysis of different biological samples. Main objective is to be able to perform the differential analysis both through DeSEQ2 and the other tool whose name I can't remember. 

## Pregunta 4 (3 puntos): 

¿Qué genes se encuentran diferencialmente expresados entre las muestras pertenecientes al grupo tratado con OHT con respecto al control tras 24h? ¿Y en el caso de las muestras tratadas con DPN, también tras 24h? Como parte de la respuesta a esta pregunta, podéis entregar una o más tablas adjuntas donde se incluyan los genes diferencialmente expresados, indicando el criterio o los criterios que habéis seguido para filtrar los resultados, así como cualquier otro gráfico o gráficos que hayáis elaborado durante el análisis.

```{r libraryload, echo = FALSE, message = FALSE}
# Library loading
library("DESeq2")
library("pheatmap")
library("RColorBrewer")
```


```{r summarizedexperiment}
# Open both tables as TSV files with header
cts <- read.table("data/input-2/rawcounts.tsv", sep = "\t", header = TRUE, row.names = 1)
metadata <- read.table("data/input-2/metadata.tsv", sep = "\t", header = TRUE, row.names = 1)

# Factorization of the parameters to overcome possible warnings and/or memory glitches
metadata$agent <- factor(metadata$agent)
metadata$patient <- factor(metadata$patient)
metadata$time <- factor(metadata$time)

# Now we'll make a new column containing a concatenation of agent and time, for resolution of #1
metadata$agent_time <- as.factor(paste(metadata$agent, metadata$time, sep="_"))

head(metadata)
all(rownames(metadata) == colnames(cts))
all(rownames(metadata) %in% colnames(cts))
```

In order to solve the first question in this second part of the exercise, first creating a subset for 24h and 48h samples, and then using this subset to pick only those counts which coincide.

```{r dds_object}
# DeSEQ2 object creation
dds <- DESeqDataSetFromMatrix(countData = cts,
                                       colData = metadata,
                                       design =~ patient + agent_time)
dds
summary(dds)
```

Pre-processing of the counts is also recommended to remove those reads with too little counts to be considered relevant. To keep consistency with how the sample experiment was performed in class with 'airway' package, we will keep the quality threshold to > 10 counts.

```{r pre_proc}
keep <- rowSums(counts(dds)) >= 10 ## Seleccionar genes con más de 10 cuentas en todos los samples
dds <- dds[keep, ]
summary(dds)
```

If we continue to follow the logic of the sample pipeline from class, a VST assay should be performed for data exploration. In my opinion, previous sub-setting of the data kind of defeats the purpose, as it is going to definitely affect data normalization and will produce biased statistics. Again, a way to overcome this problem should be approached. Then, we'll perform a distance assay between the count matrix and the variant-stabilized object.

```{r vst}
# VST object creation
vsd <- vst(dds, blind = TRUE)
```

```{r vstPCA, echo=FALSE, fig.width=6, fig.asp=1, out.width="100%", fig.align="default"}
plotPCA(vsd, intgroup = "agent")
plotPCA(vsd, intgroup = "patient")
plotPCA(vsd, intgroup = "agent_time")
```   

Distance calculation

```{r distance}
# Distance calculation
sampleDists <- dist(t(assay(vsd)))

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$agent, vsd$patient, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

After plotting PCA and Distance matrix, an obvious outlier shows up in group 4. Given that we are only going to work with 24h samples in this first part of the analysis, we can remove those from the object. Note that the removal of this outliers depends vastly on how the experiment was performed, in what we are looking to achieve, and the investigator.

```{r outlier_removal, fig.width=6, fig.asp=1, out.width="100%", fig.align="default"}
metadata = subset(metadata, patient != 4)

cts <- cts[, rownames(metadata)]
all(rownames(metadata) == colnames(cts))
all(rownames(metadata) %in% colnames(cts))

dds <- DESeqDataSetFromMatrix(countData = cts,
                                       colData = metadata,
                                       design =~ patient + agent_time)
dds
summary(dds)

keep <- rowSums(counts(dds)) >= 10 ## Seleccionar genes con más de 10 cuentas en todos los samples
dds <- dds[keep, ]
summary(dds)

vsd <- vst(dds, blind = TRUE)

plotPCA(vsd, intgroup = "agent")
plotPCA(vsd, intgroup = "patient")
plotPCA(vsd, intgroup = "agent_time")

sampleDists <- dist(t(assay(vsd)))

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$agent, vsd$patient, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

DeSEQ2 

```{r deseq}
dds2 <- DESeq(dds, test = "Wald")
plotMA(dds2)
```

Gene listing for OHT (with no log-fold threshold)

```{r results_OHT}
results_OHT <- results(object = dds2,
                      contrast = c("agent_time", "Control_24h", "OHT_24h"),
                      alpha = 0.05, # predisponer NUESTRO ERROR, por defecto tiene un 0.1
                      pAdjustMethod = "BH",
                      tidy = TRUE
                      )
mat <- assay(vsd)[head(order(results_OHT$padj), 30), ] 
pheatmap(mat)
```

Gene listing for OHT (with log-fold threshold = 1)

```{r results_threshold_OHT}
results_thresholded_OHT <- results(object = dds2,
                      contrast = c("agent_time", "Control_24h", "OHT_24h"),
                      lfcThreshold = 1,
                      alpha = 0.05,
                      pAdjustMethod = "BH",
                      tidy = TRUE
                      )
mat <- assay(vsd)[head(order(results_thresholded_OHT$padj), 30), ] 
pheatmap(mat)
```

Gene listing for DPN (with no log-fold threshold)

```{r results_DPN}
results_DPN <- results(object = dds2,
                      contrast = c("agent_time", "Control_24h", "DPN_24h"),
                      alpha = 0.05, # predisponer NUESTRO ERROR, por defecto tiene un 0.1
                      pAdjustMethod = "BH",
                      tidy = TRUE
                      )
mat <- assay(vsd)[head(order(results_DPN$padj), 30), ] 
pheatmap(mat)
```

Gene listing for DPN (with log-fold threshold = 1)

```{r results_threshold_DPN}
results_thresholded_DPN <- results(object = dds2,
                      contrast = c("agent_time", "Control_24h", "DPN_24h"),
                      lfcThreshold = 1,
                      alpha = 0.05,
                      pAdjustMethod = "BH",
                      tidy = TRUE
                      )
mat <- assay(vsd)[head(order(results_thresholded_DPN$padj), 30), ] 
pheatmap(mat)
```

## RESULTS

Given our DGE results, we can assert that, using this concatenating logic for the design/contrast of the experiment, only one DGE can be observed for DPN treatment was DNAJB9 ENSG00000128590 (ERdj4, MDG-1, MDG1, MST049, MSTP049); this number flats to ero when using a threshold = 
An isolated differential expressed gene can not be representative of any real impact of the drug in the cell expression profile, thus determining that these two treatments have no effect at 24h time. 

```{r summary_results, echo = FALSE}
paste0('No. of representative genes for DPN treatment: ', length(which(results_DPN$padj <= 0.05)))
paste0('No. of representative genes for OHT treatment: ', length(which(results_OHT$padj <= 0.05)))
paste0('No. of representative genes for DPN treatment (threshold = 1): ', length(which(results_thresholded_DPN$padj <= 0.05)))
paste0('No. of representative genes for OHT treatment (threshold = 1): ', length(which(results_thresholded_OHT$padj <= 0.05)))
```

